---
title: 'Differential Expression of Fusion Groups in Ribodepleted RNAseq'
author: "Jenny Smith"
date: "April 5, 2019"
output: html_document
---



#Set-up 

```{r setup}
library(knitr)
knitr::opts_chunk$set(tidy.opts=list(width.cutoff=50),tidy=TRUE, fig.align='center', fig.height=5, fig.width=8, dpi = 600)
knitr::opts_knit$set(root.dir = file.path(PROJHOME, '2019.04.05_Kallisto_DE_Transcriptome_MS/'))
options(stringsAsFactors = FALSE)
```

```{r message = FALSE, warning=FALSE}
library(stringr)
library(magrittr)
library(ggplot2)
library(dplyr)
library(tibble)
library(tidyr)
# library(XLConnect)
# library(DeGSEA, lib.loc = "/home/jlsmith3/R/x86_64-pc-linux-gnu-library/3.5") #my custom build R package that used to be imported through `source()` 
library(DeGSEA)
library(ggalluvial)
getwd()
```



#Define Functions 

```{r}
#Function to split the fusion columns into multiple single columns, coded as Yes/No/Unknown
createMultiCols <- function(col,split=FALSE,suffix){
  #col is the column with many factors (eg all fusions). character vector lenght 1.
  #suffix is a character string for the suffix on all column names
  #desinged for dplyr bind_cols() function
  
  #example
  # df <- df %>%
  #   bind_cols(.,createMultiCols(.$Fusion, suffix=".sensefusion")) 
  
  if(split){
    groups <- unique(gsub(" ","",unlist(str_split(col, "; "))))
    groups <- groups[groups != ""]
  }else{
    groups <- unique(col)
  }
  list <- lapply(groups, function(x) ifelse(grepl(paste0(x, ";"), col) | grepl(paste0(x, "$"), col) , "Yes",
                                            ifelse(grepl("^$|Unknown",col) | is.na(col), "Unknown","No"))) 
  list.names <- gsub("-", ".",  groups)
  names(list) <- paste0(list.names,suffix)
  return(list)
}

```

```{r}
classify_fusions <- function(FOI, ISCN){
  require(dplyr)
  require(tibble)
  
  ref <- data.frame(matrix(ncol = 3, nrow = 0,
                           dimnames = list(NULL,c("Fusion","Regex","Cyto.Regex")))) %>% 
    add_row(Fusion=c("NUP98-NSD1",
                    "NUP98-KDM5A",
                    "CBFA2T3-GLIS2",
                    "KMT2A-MLLT3", 
                    "KMT2A-MLLT10",
                    "KMT2A-MLLT4",
                    "KMT2A-ELL",
                    "KMT2A-MLLT1", 
                    "RUNX1-RUNX1T1", 
                    "CBFB-MYH11", 
                    "DEK-NUP214",
                    "FUS-ERG",
                    "KMT2A-LASP1"), 
           Regex= c("NUP98-NSD1|NSD1-NUP98", 
                    "NUP98-KDM5A|KDM5A-NUP98", 
                    "CBFA2T3-GLIS2|GLIS2-CBFA2T3", 
                    
                    "KMT2A-MLLT3|MLLT3-KMT2A", 
                    "KMT2A-MLLT10|MLLT10-KMT2A", 
                    "KMT2A-MLLT4|MLLT4-KMT2A", 
                    
                    "KMT2A-ELL|ELL-KMT2A", 
                    "KMT2A-MLLT1(;|$)|MLLT1-KMT2A(;|$)", 
                    "RUNX1-RUNX1T1|RUNX1T1-RUNX1", 
                    "CBFB-MYH11|MYH11-CBFB", 
                    
                    "DEK-NUP214|NUP214-DEK",
                    "FUS-ERG|ERG-FUS", 
                    "KMT2A-LASP1|LASP1-KMT2A"), 
           Cyto.Regex=c("t\\(\\d{0,2};?5;11;?\\d{0,2}\\)\\([pq]?.{0,};?p35;p15.{0,}\\)", 
                        "t\\(\\d{0,2};?11;12;?\\d{0,2}\\)\\([pq]?.{0,};?p15;p13.{0,}\\)", 
                        "inv\\(16\\)\\(p13;?q24\\)", 
                        
                        "t\\(\\d{0,2};?9;11;?\\d{0,2}\\)\\s?\\([pq]?.{0,};?p2\\d;q23.{0,}\\)", 
                        "(t|dic|ins)\\(\\d{0,2};?10;11;?\\d{0,2}\\)\\([pq]?.{0,};?p1\\d;q23.{0,}\\)",  
                        "t\\(\\d{0,2};?6;11;?\\d{0,2}\\)\\([pq]?.{0,};?q2\\d;q23.{0,}\\)", 
                        
                        "t\\(\\d{0,2};?11;19;?\\d{0,2}\\)\\([pq]?.{0,};?q23;?p13.1.{0,}\\)", 
                        "t\\(\\d{0,2};?11;19;?\\d{0,2}\\)\\([pq]?.{0,};?q23;?p13.[23].{0,}\\)", 
                        "t\\(\\d{0,2};?8;21;?\\d{0,2}\\)\\(q22;?q22\\)|ins\\(21;8\\)|ins\\(8;21\\)", 
                        "inv\\(16\\)\\(p13.{0,3}q22\\)|t\\(16;16\\)\\(p13.{0,3}q22|ins\\(16;16\\)\\(p13.1;?q22", 
                        
                        "t\\(\\d{0,2};?6;9;?\\d{0,2}\\)\\([pq]?.{0,};?p2[23];?[pq]{0,1}.{0,3};q34.{0,}\\)|t\\(\\d{0,2};?9;6;?\\d{0,2}\\)\\([pq]?.{0,};?q34;p2[23].{0,}\\)", 
                        "t?\\(\\d{0,2};?16;21;?\\d{0,2}\\)\\([pq]?.{0,};?p11.{0,2};q22.{0,}\\)",     
                        "t\\(\\d{0,2};?11;17;?\\d{0,2}\\)\\([pq]?.{0,};?q23;q[1-2][1-2].{0,}\\)")) %>% 
        set_rownames(.$Fusion)
  
  res <- NULL
  for(r in 1:nrow(ref)){
    df <- ref[r, ]
  
    #Data Query
    FOI.Present <- any(grepl(df$Regex,FOI))
    Cyto.Present <- grepl(df$Cyto.Regex, ISCN)
    No.Cyto <- is.na(ISCN)

    #define classifications
    if(FOI.Present & No.Cyto){
      res <- c(res,"RNA seq only; no cyto avail")
    }else if(FOI.Present & Cyto.Present){
      res <- c(res,"both confirms")
    } else if(FOI.Present & !Cyto.Present){
      res <- c(res,"RNA seq only")
    } else if(!FOI.Present & Cyto.Present){
      res <- c(res,"cyto only")
    }else if(!FOI.Present & ! Cyto.Present){
      res <- c(res,NA)
    }
  }

 names(res) <- ref[,"Fusion"]
 if(sum(!is.na(res)) == 0){
   res <- NA 
   names(res) <- "none"
 }else if(sum(!is.na(res)) == 1 ){
   res <- res[! is.na(res)]
   
 }else if(sum(!is.na(res)) > 1){
   n <- paste(names(res[! is.na(res)]),collapse = "; ")
   res <- "fusion/cyto data conflict"
   names(res) <- n
 }
  
  res <- paste(names(res), res, sep=": ")
  return(res)
}
```


```{r}
rmDups <- function(count.matrix, ID.map, matrix.class=TRUE,rowname.GeneSym=TRUE){
  
  df <- count.matrix %>%
    as.data.frame() %>%
    rownames_to_column("gene_id") %>%
    left_join(., ID.map, by=c("gene_id")) %>% 
    select(gene_id, gene_name, everything()) %>%
    filter(!grepl("_PAR_", gene_id)) #specific to gencode refs
  
  
  dup <- df$gene_name[which(duplicated(df$gene_name))]
  ddf <- df %>%
    filter(gene_name %in% dup) %>% 
    arrange(gene_name)%>% 
    mutate(Variance=genefilter::rowVars(select(., -gene_id, -gene_name))) %>%

    group_by(gene_name) %>% 
    mutate(High.Variance=Variance==max(Variance)) %>% 
    ungroup() %>% 
    
    filter(High.Variance)  %>%
    filter(!duplicated(gene_name))#if variance tied, pick first one
    
  
  if(rowname.GeneSym){
    rmDups.df <- df %>% 
        filter(! gene_name %in% dup) %>% 
        bind_rows(.,ddf) %>%
        select(everything(),-gene_id, -Variance,-High.Variance) %>%
        column_to_rownames("gene_name")
  }else{
    rmDups.df <- df %>% 
        filter(! gene_name %in% dup) %>% 
        bind_rows(.,ddf) %>%
        select(everything(),-gene_name, -Variance,-High.Variance) %>%
        mutate(gene_id=gsub("\\.[0-9]{1,2}", "", gene_id)) %>% 
        column_to_rownames("gene_id")
  }

  if(matrix.class){
    rmDups.df <- data.matrix(rmDups.df)
  }
  
  return(rmDups.df)
}  
```


#Read in the Clinical Data

```{r}
# merged <- read.csv(file.path(REFS,"00_Old","TARGET_AML_0531_1031_merged_CDEs_3.01.19.csv"))
# merged <- read.csv(file.path(REFS,"TARGET_AML_0531_1031_merged_CDEs_9.4.19.csv"))
merged <- read.csv(file.path(CDE,"Merged/TARGET_AML_0531_1031_merged_CDEs_7.08.20.csv"))


merged <- merged %>% 
  filter(!is.na(USI)) %>% 
  set_rownames(.$USI)

# head(merged[,1:5])
dim(merged)
```

```{r}
manifest <- read.csv(file.path(TARGET,"SequencingDataMatrix/00_archive/TARGET_AML_RBD_0531_1031_miRNAseq_mRNAseq_Manifest_v5.csv"))

# head(manifest)
dim(manifest)
```

```{r}
primary.fusion <- read.csv(file.path(HOME, "2019.04.22_Transcriptome_Profile_AML/Final_list_AAML1031_fusions_with_cyto_information.csv"))

# head(primary.fusion)
dim(primary.fusion)
```

```{r}
table(primary.fusion$Primary.Fusion.CNV) %>% 
  as.data.frame() %>% 
  arrange(desc(Freq)) %>% 
  head()
```

```{r}
all_fusions <- read.csv(file.path(TARGET, "RNA/mRNAseq/level2/fusion/concat_matrices/2017July_BCCA_0531_1031_TransAbyss_STAR_K.Mungall_Combined_Illumina_data/TARGET_AML_0531_1031_TransAbyss_STAR_TargetedAlignment_Condensed_Fusion_Columns_4.15.19.csv")) %>% 
  mutate(USI=str_split_fixed(Patient, "\\.", n=5)[,3])

# head(all_fusions)
dim(all_fusions)
```


#Define Fusion Groups on RNA-seq and Cytogenetic information

1. NUP98-NSD1 t(5;11)(q35;p15.5) (PacBio)
2. NUP98-KDM5A t(11;12)(p15;p13) 
3. CBFA2T3-GLIS2 inv(16)(p13q24) (PacBio)
MLL  11q23
  4. KMT2A-MLLT3 t(9;11)(p21;q23) (PacBio)
  5. KMT2A-MLLT10 t(10;11)(p12;q23)
  6. KMT2A-MLLT4 t(6;11)(q27;q23) 
  7. KMT2A-ELL t(11;19)(q23;p13.1)
  8. KMT2A-MLLT1 t(11;19)(q23;p13.3) 
9. RUNX1-RUNX1T1 t(8;21) (PacBio)
10.CBFB-MYH11 inv(16)

*additional_not for DE* 
11. DEK-NUP214 t(6;9)(p22;q34)
12. FUS-ERG  t(16;21)(p11;q22) 
13. KMT2A-LASP1 t(11;17)(q23;q12-21)


```{r warning=FALSE}
fusions.info <- manifest %>% 
  
  #Select columns and clean 
  select(PATIENT_ID_Original, Final_Patient_ID, 
         USI, Reg.,Replicate,Type, Batch,LIBRARY.mRNA) %>% 
  filter(Type != "CellLine", !is.na(LIBRARY.mRNA)) %>%
  
  #Add replicate information 
  mutate(Replicate=ifelse(grepl("Rep", Replicate), "_replicate", "")) %>% 
  unite(col = "Colname",Final_Patient_ID, Replicate, sep = "", remove = FALSE ) %>% 
  filter(!grepl("PATGIG.+rep|PATISD.+rep", Colname)) %>% #duplicates 
  
  #Make Ptient barcodes the same
  mutate_all(funs(gsub("-", "\\.", .))) %>% 
  mutate(Reg.=as.character(Reg.)) %>%
  
  #Add in curated fusion info from ISCN and RNAseq by Rhonda
  left_join(.,select(primary.fusion,Reg.=Patient.registration.number,
                     Primary.Fusion.CNV, Cyto.vs..Seq,
                     Additional.Fusions.CNV ,Comments, 
                     Protocol=Study, ISCN) %>%
              mutate(Reg.=as.character(Reg.)) %>% #Study, ISCN,
              filter( ! is.na(Reg.)), #filter only remove Stella to avoid duplicates
            by=c("Reg.")) %>%
  mutate_at(vars(Primary.Fusion.CNV, Cyto.vs..Seq), funs(gsub(" $", "", .))) %>%
  mutate(Reg.=ifelse(is.na(Reg.), USI, Reg.)) %>%
  
  #Add in the RNA-seq FOI information
  left_join(., select(all_fusions, Patient, FOI=FOI), 
            by=c("Colname"="Patient")) %>%
  
  #Add in ISCN and Study 
  left_join(., select(merged, Reg., Protocol,ISCN, 
                      Known.Fusion.detected.by.any.method) %>% 
              mutate(Reg.=as.character(Reg.)),
            by="Reg.") %>% 
  mutate(ISCN=ifelse(is.na(ISCN.x), ISCN.y, ISCN.x)) %>% 
  mutate(ISCN=ifelse(grepl("Unknown", ISCN) | ISCN == "", NA,  ISCN)) %>%
  mutate(Protocol=ifelse(is.na(Protocol.x), Protocol.y, Protocol.x)) %>% 
  select(-ISCN.x, -ISCN.y, -Protocol.x, -Protocol.y) 


# head(fusions.info)
dim(fusions.info) #1568   16
```

```{r warning=FALSE}
fusionGroup.info <- fusions.info %>% 
  
  dplyr::rowwise() %>%
  mutate(Cyto.vs..Seq_JSmith=classify_fusions(FOI=FOI, ISCN=ISCN)) %>%
  ungroup() %>%
  
  #keep only groups of interest for DE analysis, all others considered negative in pheno vector.
  filter(!grepl("none",Cyto.vs..Seq_JSmith)) %>% 
  
  separate(Cyto.vs..Seq_JSmith,
           into = c("Primary.Fusion_JSmith","Cyto.vs..Seq_JSmith"), sep=": ") %>% 
  mutate(Different=ifelse(!is.na(Cyto.vs..Seq) & Cyto.vs..Seq != Cyto.vs..Seq_JSmith, 
                          "Different", "Same")) %>%
  mutate(Cyto.vs..Seq_forPlot=case_when(
    grepl("PARVSF|PAYIET", USI) ~ Cyto.vs..Seq_JSmith, #found the ISCN or found batch 2 fusions calls so different than Rhonda 
    Different == "Different" & Primary.Fusion_JSmith == Primary.Fusion.CNV ~ Cyto.vs..Seq, 
    Different == "Different" & Primary.Fusion_JSmith != Primary.Fusion.CNV ~ Cyto.vs..Seq_JSmith, #I did't have monosomy7 or del5q take precedence
    is.na(Primary.Fusion.CNV) ~  Cyto.vs..Seq_JSmith,
    Different == "Same" & Primary.Fusion_JSmith == Primary.Fusion.CNV ~  Cyto.vs..Seq_JSmith,
    TRUE ~ "WTH")) %>% 
  filter(USI != "PAXDBF" & USI != "PAUPYX" & USI != "PAVNUJ") %>% #this is KMT2A-MLLT6 by RNAseq. Cyto fits the KMT2A-LASP1, but both genes fall in same bands. The last two are KMT2A-LASP1 by cyto only again. with different classifcations by Rhonda. 

  #Reorder and arrange columns 
  select(1:8, matches("Primary.Fusion|Cyto.vs"),Different, everything(), -Replicate) %>% 
  arrange(Batch,Primary.Fusion_JSmith,Cyto.vs..Seq_JSmith) %>% 
  
  #I will avoid making executive decision here. will update once Rhonda/Jason Farrar can decide. 
  filter(Cyto.vs..Seq_JSmith != "fusion/cyto data conflict") %>% 
  bind_cols(.,createMultiCols(.$Primary.Fusion_JSmith, suffix=".forDEGs")) #%>% 
  # filter(duplicated(Final_Patient_ID) | duplicated(Final_Patient_ID, fromLast=TRUE))


head(fusionGroup.info)
dim(fusionGroup.info) #807  32
```

```{r}
# table(fusionGroup.info$Batch) #565 in batch 1 and 197  in batch 2
table(fusionGroup.info$Cyto.vs..Seq_JSmith,fusionGroup.info$Different)
table(fusionGroup.info$Primary.Fusion_JSmith, useNA = "always")
```

```{r}
# write.csv(fusionGroup.info, "TARGET_AML_RBD_Fusion_Cyto_vs_RNAseq_GroupsOfInterest_.csv", row.names = FALSE)
```


```{r}
temp <- fusionGroup.info %>% 
  filter(grepl("ERG|DEK|LASP1", Primary.Fusion_JSmith)) 
  # select(USI,Primary.Fusion_JSmith,FOI, ISCN,Cyto.vs..Seq_JSmith)
```


table(fusion.info$Cyto.vs..Seq, useNA = "always")
table(fusion.info$Primary.Fusion.CNV, useNA = "always")
table(fusion.info$RNASeq_FOI, useNA = "always")
table(filter(fusion.info, grepl("KMT2A", Primary.Fusion.CNV)) %>% select(Primary.Fusion.CNV), useNA = "always") %>% as.data.frame() %>% filter(Freq >= 3) %>% arrange(desc(Freq))

table(filter(fusion.info, grepl("NSD1", RNASeq_FOI)) %>% select(RNASeq_FOI), useNA = "always") %>% as.data.frame() #%>% filter(Freq >= 3) %>% arrange(desc(Freq))  
table(filter(fusion.info, grepl("KDM5A", RNASeq_FOI)) %>% select(RNASeq_FOI), useNA = "always") %>% as.data.frame() #%>% filter(Freq >= 3) %>% arrange(desc(Freq))  

##Alluvial Plot 

```{r}
colors <- c("#E41A1C", "#377EB8" ,"#4DAF4A" ,
                 "#F781BF","blue1", "darkslategray3","burlywood3", "#984EA3", "#FF7F00",
                "seagreen2", "maroon", "orchid", "cornflowerblue", "yellow2",
                "darkblue", "azure4", "chartreuse1", "orange1",
                "deeppink", "darkslategray1", "green4", "navajowhite2",
                "brown3", "darkgoldenrod3", "deepskyblue1", "lightcoral",
                "mediumorchid", "darkmagenta")
```


```{r fig.width=12, fig.height=10, warning=FALSE}
ggDat_A <- fusionGroup.info %>%
  filter(Type=="AML", !grepl("replicate", Colname)) %>% 
  select(USI,Fusion=Primary.Fusion_JSmith, Detection=Cyto.vs..Seq_JSmith) %>% 
  filter(! grepl("NUP214|ERG|LASP1",Fusion)) %>% #704 patients included here
  mutate(Cytogenetics=ifelse(grepl("both confirms|cyto only",Detection), "Yes","No"), 
         RNAseq=case_when(
           grepl("no cyto avail", Detection)~ "Yes\n(No Cyto Avail)",
           grepl("both confirms|RNA seq only", Detection)~ "Yes",
           TRUE~ "No")) %>% 
  group_by(Fusion,Cytogenetics,RNAseq) %>%
  summarize(Freq=n()) %>%
  ungroup() %>% 
  arrange(desc(Cytogenetics)) %>%
  mutate(Fusion=factor(Fusion, levels = unique(Fusion)), 
         Cytogenetics=factor(Cytogenetics, levels = c("Yes","No")), 
         RNAseq=factor(RNAseq, levels = c("Yes","Yes\n(No Cyto Avail)","No")))

w <- 1/2
alluvial_A <- ggplot(data= ggDat_A, aes(axis1=Fusion, 
                                    axis2=Cytogenetics,
                                    axis3=RNAseq,
                                    y=Freq)) +
  scale_x_discrete(limits=c("Fusion","Cytogenetics","RNAseq"), 
                   labels=c("Fusion","Cytogenetics\nDetection","RNAseq\nDetection"),
                   expand = c(0.001,0.1,0.1,0.001)) + 
  scale_y_continuous(breaks = seq(0,705,by=100)) +
  geom_alluvium(aes(fill=Fusion), alpha=0.8, width = w) +
  geom_stratum(alpha=0.7, width = w) +
  
  ggfittext::geom_fit_text(data= ggDat_A, aes(axis1=Fusion, 
                                    y=Freq), #Ignoring unknown aesthetics: axis1 ???
                           stat = "stratum",fontface = "bold",
                           grow=TRUE,reflow = TRUE,
                           width = w, min.size = 8, 
                           label.strata=TRUE, 
                           inherit.aes = FALSE) +
  
   ggfittext::geom_fit_text(data= mutate(ggDat_A,Temp=""), 
                            aes(axis1=Temp,axis2=Cytogenetics, axis3=RNAseq, y=Freq), #Ignoring unknown aesthetics: axis1, axis2 ???
                           stat = "stratum",size=26,
                           grow=FALSE,reflow = FALSE,
                           width = w, min.size = 10, 
                           label.strata=TRUE, 
                           inherit.aes = FALSE) +
  
  theme_minimal() +
  ggtitle("AML Fusion Subtypes Investigated") + 
  labs(y="Number of Patients",x="") +
  theme(plot.title = element_text(size=30,face="bold"),
        panel.grid.major = element_line(color="dark grey"),
        axis.title.y= element_text(size=26),
        axis.text = element_text(size=20, color="black"),
        legend.title = element_blank(),
        legend.text = element_text(size=16), 
        legend.position = "top")

# alluvial_A
# ggsave(plot=alluvial_A, filename="Fusion_Detection_Alluvial_plot.png", device = "png", dpi=350,units="in", height = 10, width = 12)
# saveRDS(alluvial_A,file="Fusion_Detection_Alluvial_plot.RDS")
```

```{r}
ggDat_A %>% 
  filter(grepl("Yes", RNAseq)) %>% 
  mutate(N=sum(Freq))
```

```{r}
30/705*100 #4.36% error rate
674/705*100 #95.74% sensitivity 
705/1394*100 #50.6% of patients investigated
```


```{r fig.height=10, fig.width=10}
ggDat_B <- fusionGroup.info %>%
  filter(Type=="AML", !grepl("replicate", Colname)) %>% 
  select(USI,Fusion=Primary.Fusion_JSmith, Detection=Cyto.vs..Seq_JSmith) %>% 
  filter(! grepl("NUP214|ERG|LASP1",Fusion)) %>% 
  group_by(Fusion, Detection) %>%
  summarise(Freq=n()) %>%
  ungroup() %>%
  arrange(Detection) %>%
  mutate(Fusion=factor(Fusion, levels = unique(Fusion)),
         Detection=gsub("; ", "\n", Detection))


alluvial_B <- ggplot(data= ggDat_B, aes(axis1=Fusion, 
                                    axis2=Detection, 
                                    y=Freq)) +
  scale_x_discrete(limits=c("Fusion","Detection"), 
                   expand = c(0.001,0.001)) + 
  geom_alluvium(aes(fill=Fusion), alpha=0.8) +
  geom_stratum(alpha=0.5) +
  ggfittext::geom_fit_text(stat = "stratum", 
                           width = 3/4, min.size = 5, 
                           label.strata=TRUE) +
  # geom_text(stat = "stratum", label.strata = TRUE) +
  theme_minimal() +
  ggtitle("AML Fusion Subtypes Investigated") 

# alluvial_B
```



#Read in the Gene Annotation References

https://bioconductor.org/packages/release/bioc/vignettes/tximport/inst/doc/tximport.html#limma-voom

```{r}
anno <- read.csv(file.path(HOME, "0000.00.02_Reference_GeneInfo/gencode.v29_RepBase.v24.01_GeneLevel_IDmap_anno.csv"))

head(anno)
dim(anno)
```

```{r}
lncRNA <- read.csv(file.path(HOME, "0000.00.02_Reference_GeneInfo/gencode.v29.lncRNAs_Transcript.IDmap.csv")) %>%
  select(gene_id, gene_type, gene_name) %>%
  mutate_at(vars(gene_id),
            funs(gsub("\\.[0-9]{1,}$", "", .))) %>%
  unique()

head(lncRNA)
dim(lncRNA) #16066  lncRNA genes   3

# write.csv(lncRNA, "~/RNA_seq_Analysis/0000.00.02_Reference_GeneInfo/gencode.v29.lncRNAs_GeneLevel.IDmap.csv")
```

```{r}
table(lncRNA$gene_type)
```


#Read in the counts

```{r}
cts <- readRDS(file.path(HOME, "0000.00.03_Expression_Matrices/TARGET_AML_RBD_Kallisto_Quant_GeneLevel_scaledTPM_counts.RDS"))

#remove version numbers from the counts
rownames(cts) <- gsub("\\.[0-9]{1,}$", "", rownames(cts))


head(cts[,1:5])
dim(cts)
```

```{r}
TPMs <- readRDS(file.path(HOME,"0000.00.03_Expression_Matrices/TARGET_AML_RBD_Kallisto_Quant_GeneLevel_TPM.RDS"))

rownames(TPMs) <- gsub("\\.[0-9]{1,}$", "", rownames(TPMs))
head(TPMs[,1:5])
dim(TPMs) #59853  1462
```


#Define Cohorts for DE Analysis 

```{r}
# fusionGroup.info <- read.csv("TARGET_AML_RBD_Fusion_Cyto_vs_RNAseq_GroupsOfInterest_.csv")
cols <- c("Cytogenetic.Category.1", "Cytogenetic.Category.2", "SNVs","Rare.Fusions")

DE.Groups <- fusionGroup.info %>% 
  
  #select only bulk AMLs initialy
  filter(Type=="AML" & ! grepl("replicate", Colname)) %>%
  select(USI,contains(".forDEGs")) %>%
  
  #Supply remaining USIs
  right_join(., data.frame(USI=colnames(cts)),
            by="USI") %>% 
  filter(!grepl("^BM|^RO", USI)) %>%
  mutate(AML="AML") %>% 
  
  #Add in FOI columns
  left_join(., fusions.info %>% 
              filter(Type=="AML" & ! grepl("replicate", Colname)) %>%
              select(USI,Reg.,FOI), 
            by="USI") %>%
  
  #Add in heatmap columns
  left_join(., select(merged,Reg.,cols) %>% 
               mutate(Reg.=as.character(Reg.)),
            by="Reg.") %>%
  
  mutate_at(vars(contains(".forDEGs")), funs(case_when(
    is.na(.) ~ "No",
    TRUE ~ .))) %>% 
  
 mutate_at(vars( "Cytogenetic.Category.2", "SNVs"),
           funs(case_when(
                  is.na(.) ~ "Unknown",
                  TRUE ~ .))) %>% 
  
  mutate_at(vars("Rare.Fusions"),
            funs(case_when(
            (. == "OtherAML" | is.na(.)) & grepl("CBFA2T3.GLIS2", FOI) ~ "CBFA2T3.GLIS2",
            (. == "OtherAML" | is.na(.)) & grepl("NSD1", FOI) ~ "NUP98.NSD1",
            (. == "OtherAML" | is.na(.)) & grepl("DEK.NUP214", FOI) ~ "DEK.NUP214",
            (. == "OtherAML" | is.na(.)) & grepl("NPM1.MLF1", FOI) ~ "NPM1.MLF1",
            (. == "OtherAML" | is.na(.)) & grepl("KDM5A", FOI) ~ "NUP98.KDM5A",
            (. == "OtherAML" | is.na(.)) & grepl("RBM15.MKL1", FOI) ~ "RBM15.MKL1",
            (. == "OtherAML" | is.na(.)) & grepl("RUNX1.CBFA2T3", FOI) ~ "RUNX1.CBFA2T3",
            NUP98.KDM5A.forDEGs == "No" & grepl("NUP98.KDM5A",.) ~ "OtherAML", #incorrect classification
            NUP98.NSD1.forDEGs == "No" & grepl("NUP98.NSD1",.) ~ "OtherAML",
            is.na(.) ~ "Unknown",
            TRUE ~ .))) %>%
  
  mutate_at(vars(Cytogenetic.Category.1),
            funs(case_when(
              is.na(.) & grepl("KMT2A", FOI) ~ "MLL", 
              is.na(.) & grepl("CBFB.MYH11", FOI) ~ "inv.16", 
              is.na(.) & grepl("RUNX1.RUNX1T1", FOI) ~ "t.8.21.",
              is.na(.) ~ "Unknown", 
              TRUE ~ .))) %>% 
  
  select(USI, Reg. ,FOI,AML, everything()) %>%
  as.data.frame() %>%
  set_rownames(.$USI)
  
# head(DE.Groups)  
dim(DE.Groups) #1394   13
# write.csv(DE.Groups, "TARGET_AML_RBD_Cohorts_forDE_5.20.19.csv", row.names = FALSE)
```

```{r eval=FALSE}
DE.Groups <- read.csv("TARGET_AML_RBD_Cohorts_forDE_5.20.19.csv") %>% 
  set_rownames(.$USI)

# sum(is.na(DE.Groups$Reg.))
sapply(DE.Groups[,grep("forDEGs", colnames(DE.Groups))], table, useNA="always")

head(DE.Groups)
dim(DE.Groups)
# table(DE.Groups$Cytogenetic.Category.1, useNA = "always")
# table(DE.Groups$Rare.Fusions, useNA = "always")
```


#Differential Expression Analysis

```{r}
library(rslurm)
```

```{r}
sopt <- list('nodes'='1', 'cpus-per-task'='2',
             'partition'='campus-new', 'mem'='14G',
             'time' = '24:00:00', 'mail-type'='FAIL,END',
             'mail-user'='jlsmith3@fredhutch.org') 

```

```{r}
ID.map <- "~/RNA_seq_Analysis/0000.00.02_Reference_GeneInfo/gencode.v29_RepBase.v24.01_GeneLevel_IDmap.csv"

head(ID.map)
```


## AML vs NBM 

```{r warning=FALSE, eval=FALSE}
path="/fh/scratch/delete90/meshinchi_s/jlsmith3/kallisto"
setwd(path)
NBM_all <- slurm_call(f=twoGroups_DEGs,
                     jobname = "AMLvsNBM_allGenes",
                     params =  list(expnData = cts, 
                                    clinData = DE.Groups, 
                                    col="AML", 
                                    ref="No", 
                                    anno=TRUE,
                                    ids2symbols=ID.map,
                                    gene.name.col="gene_name",
                                    BM=TRUE),
                     add_objects = c("cts","DE.Groups"),
                     slurm_options=sopt,
                     submit = TRUE) #Submitted batch job 

print_job_status(NBM_all)
```

```{r warning=FALSE, eval=FALSE}
path="/fh/scratch/delete90/meshinchi_s/jlsmith3/kallisto"
setwd(path)
NBM_lncs <- slurm_call(f=twoGroups_DEGs,
                     jobname = "AMLvsNBM_lncRNA",
                     params =  list(expnData = cts, 
                                    clinData = DE.Groups, 
                                    col="AML", 
                                    ref="No",
                                    anno=FALSE,
                                    GOI=lncRNA[["gene_id"]],
                                    BM=TRUE),
                     add_objects = c("cts","DE.Groups"),
                     slurm_options=sopt,
                     submit = TRUE) #Submitted batch job 

print_job_status(NBM_lncs)
```



## Pairwise Fusion Groups 

```{r warning=FALSE, eval=FALSE}
path="/fh/scratch/delete90/meshinchi_s/jlsmith3/kallisto"
setwd(path)

cols <- grep("forDEGs", colnames(DE.Groups), value=TRUE)
DEGs_NBM <- lapply(1:10, function(i) slurm_call(f=twoGroups_DEGs,
                     jobname = paste0(gsub("forDEGs","", cols[i]),"vsNBM"),
                     params =  list(expnData = cts, 
                                    clinData = DE.Groups, 
                                    col=cols[i], 
                                    ref="No", 
                                    anno=FALSE,
                                    ids2symbols=ID.map,
                                    gene.name.col="gene_name",
                                    BM=TRUE),
                     add_objects = c("cts","DE.Groups"),
                     slurm_options=sopt,
                     submit = TRUE)) #Submitted batch job 


```

```{r warning=FALSE, eval=FALSE}
path="/fh/scratch/delete90/meshinchi_s/jlsmith3/kallisto"
setwd(path)

cols <- grep("forDEGs", colnames(DE.Groups), value=TRUE)
DEGs_all <- lapply(1:10, function(i) slurm_call(f=twoGroups_DEGs,
                     jobname = paste0(gsub("forDEGs","", cols[i]),"vsOthers"),
                     params =  list(expnData = cts, 
                                    clinData = DE.Groups, 
                                    col=cols[i], 
                                    ref="No", 
                                    anno=TRUE,
                                    ids2symbols=ID.map,
                                    gene.name.col="gene_name",
                                    BM=FALSE),
                     add_objects = c("cts","DE.Groups"),
                     slurm_options=sopt,
                     submit = TRUE)) #Submitted batch job 


```

```{r warning=FALSE, eval=FALSE}
path="/fh/scratch/delete90/meshinchi_s/jlsmith3/kallisto"
setwd(path)

cols <- grep("forDEGs", colnames(DE.Groups), value=TRUE)
DEGs_lncs_NBM <- lapply(1:10, function(i) slurm_call(f=twoGroups_DEGs,
                     jobname = paste0(gsub("forDEGs","", cols[i]),"vsNBM_lncRNA"),
                     params =  list(expnData = cts, 
                                    clinData = DE.Groups, 
                                    col=cols[i], 
                                    ref="No",
                                    anno=FALSE,
                                    GOI=lncRNA[["gene_id"]],
                                    BM=TRUE),
                     add_objects = c("cts","DE.Groups"),
                     slurm_options=sopt,
                     submit = TRUE)) #Submitted batch job 
```

```{r warning=FALSE, eval=FALSE}
path="/fh/scratch/delete90/meshinchi_s/jlsmith3/kallisto"
setwd(path)

cols <- grep("forDEGs", colnames(DE.Groups), value=TRUE)
DEGs_lncs <- lapply(1:10, function(i) slurm_call(f=twoGroups_DEGs,
                     jobname = paste0(gsub("forDEGs","", cols[i]),"vsOthers_lncRNA"),
                     params =  list(expnData = cts, 
                                    clinData = DE.Groups, 
                                    col=cols[i], 
                                    ref="No",
                                    anno=FALSE,
                                    GOI=lncRNA[["gene_id"]],
                                    BM=FALSE),
                     add_objects = c("cts","DE.Groups"),
                     slurm_options=sopt,
                     submit = TRUE)) #Submitted batch job 
```


### For Ben Huang

I annoyingly deleted the Rdata objects to save disk space.... 
But I still use these results so that was not a great move. 


```{r}
conversion <- read.delim(file.path(HOME,"0000.00.02_Reference_GeneInfo/gencode.v29_RepBase.v24.01_GeneLevel_IDmap_anno_5.14.21.txt"))

dim(conversion)
```

```{r}
DEGs.CBFGLIS.orig <- read.csv("DEGs/CBFA2T3GLIS2vsNBM_DEGs.csv")
dim(DEGs.CBFGLIS.orig)
head(DEGs.CBFGLIS.orig)


CBFGLISvsNBM <- read.csv(file.path(PROJHOME, "0000.00.05_DEG_Lists/GRCh38_hg38/Kallisto/CBFA2T3GLIS2vsNBM_DEGs.csv"))

dim(CBFGLISvsNBM)
head(CBFGLISvsNBM, n=10)

# identical(DEGs.CBFGLIS.orig$gene, CBFGLISvsNBM$gene)
# identical(DEGs.CBFGLIS.orig$logFC, CBFGLISvsNBM$logFC)
```

```{r}
DEGs.CBFGLIS <- twoGroups_DEGs(expnData = cts, 
                               clinData = DE.Groups, 
                               col = "CBFA2T3.GLIS2.forDEGs",
                               ref="No",
                               BM = TRUE,
                               anno = FALSE,
                               SkipPlots = TRUE,
                               percent.cutoff = 0.05)
```

```{r}
table(DEGs.CBFGLIS$phenovector)
temp <- extract_DEGs(DEGs.CBFGLIS)
dim(temp)

#Identical
all(temp$gene %in% DEGs.CBFGLIS.orig$gene)
all(DEGs.CBFGLIS.orig$gene %in% temp$gene)
# bind_cols(tibble(Orig=DEGs.CBFGLIS.orig$logFC, 
#                   Redo=temp$logFC)) %>% 
#   mutate_all(~as.character(.)) 

# class(DEGs.CBFGLIS.orig$logFC)
# class(temp$logFC)
```


```{r}
  fit <- DEGs.CBFGLIS$DE$eBayesFit
  df <- data.frame(gene=rownames(fit[["coefficients"]]),
              log2FC=fit[["coefficients"]][,1],
              p.value=fit[["p.value"]][,1]) %>%
    mutate(FDR=p.adjust(p.value, method = "BH")) %>%
    left_join(., select(conversion, gene_id, gene_name), 
              by=c("gene"="gene_id")) %>% 
    arrange(desc(log2FC), desc(FDR)) %>% 
    select(gene, gene_name, everything())

  
head(df)
dim(df) #27541     4
  # filename <- file.path(PROJHOME, "2017.02.15_CBF-GLIS_DEG/2020.12.23_CBFGLIS_Models", paste0("DEGs/for_volcano_plots/TARGET_AML_CBFGLIS_vs_NBM_DEGs_forVolcanoPlots.csv"))
  # con <- file(filename, open="wt")
  # writeLines(paste("# Differentially expressed genes in CBFGLIS patient samples compared to Normal bone marrow (NBM) healthy controls"), con)
  # writeLines(paste("# Input Samples: CBFGLIS bulk RNAseq (N=39) vs  NBM bulk RNAseq (N=68)"), con)
  # write.csv(df, con, row.names = F)
  # close(con)
```


# Read in the DE Results

```{r}
path="/fh/scratch/delete90/meshinchi_s/jlsmith3/kallisto"
files <- dir(path=path, pattern="results_0.RDS", recursive=TRUE, full.names = TRUE) %>% 
  .[!grepl("tximport|old", .)] #%>% 
  # grep("vsOthers_lncRNA", ., value=TRUE)

files
```


```{r}
AMLvsNBM <- readRDS("/fh/scratch/delete90/meshinchi_s/jlsmith3/kallisto/_rslurm_AMLvsNBM_lncRNA/results_0.RDS")
```

```{r}
for (i in 1:length(files)){
  
  name <- gsub("^.+rslurm_(.+)\\/.+", "\\1", files[i])
  
  filename <- paste0(getwd(),"/") %>%
    paste0(., name) %>% 
    paste0(.,".RDS")
  
  print(filename)
  print(name)
  assign(name, readRDS(files[i]))
  
  # file.copy(files[i], filename, overwrite = FALSE)
}
```



#Extract DEGs 

```{r}
# Others_lncRNA <- dir(pattern = "Others_lncRNA.RDS", include.dirs = TRUE, recursive = TRUE)
# # Others_lncRNA
# 
# for(i in 1:length(Others_lncRNA)){
#   name <- gsub("^.+\\/(.+)\\.RDS", "\\1", Others_lncRNA[i])
#   print(name)
#   assign(name, readRDS(Others_lncRNA[i]))
# }

# AML.Others <- dir(pattern = "[A-Z0-9]vsOthers.RDS",
#                   include.dirs = TRUE, 
#                   recursive = TRUE)
# 
# for(i in 1:length(AML.Others)){
#   name <- gsub("^.+\\/(.+)\\.RDS", "\\1", AML.Others[i])
#   print(name)
#   assign(name, readRDS(AML.Others[i]))
# }

# AML.NBM <- dir(pattern = "vsNBM.RDS|AMLvsNBM_allGenes.RDS", include.dirs = TRUE, recursive = TRUE)
# # AML.NBM
# 
# for(i in 1:length(AML.NBM)){
#   name <- gsub("^.+\\/(.+)\\.RDS", "\\1", AML.NBM[i])
#   print(name)
#   assign(name, readRDS(AML.NBM[i]))
# }
```

Had java memory issues with XLConnect package... 

```{r eval=FALSE}
results <- ls(pattern = "*vs*")
for (res in results){
  print(res)
  
  if(grepl("vsOthers$", res)){
    df <- extract_DEGs(get(res), anno = TRUE, geneLevel = TRUE)
    write.csv(df, paste0("DEGs/", res, "_DEGs.csv"), row.names=FALSE)

    # createSheet(others, name = res )
    # writeWorksheet(others, df, sheet = res)
  }else if (grepl("vsNBM$|AMLvsNBM_allGenes", res)){
    df <- extract_DEGs(get(res), anno = TRUE, geneLevel = TRUE)
    write.csv(df, paste0("DEGs/", res, "_DEGs.csv"), row.names=FALSE)

    # createSheet(NBM, name = res)
    # writeWorksheet(NBM, extract_DEGs(get(res), anno = TRUE, geneLevel = TRUE), sheet = res)
  }else if(grepl("lncRNA", res)){
    
    df <- extract_DEGs(get(res)) %>%
      left_join(.,  lncRNA, by=c("gene"="gene_id"))
    
    write.csv(df, paste0("DEGs/", res, "_DEGs.csv"), row.names=FALSE)

    # createSheet(lnc, name = res)
    # writeWorksheet(lnc, df, sheet = res)
  }
}

```


# Save Plots 

```{r}
results <- ls(pattern = ".+vs")
results
```

```{r}
plots_MDS_PCA <- list()
for (res in results){
  plots_MDS_PCA[[paste0(res,".PCA")]] <- extract_PCA(get(res))
  plots_MDS_PCA[[paste0(res,".MDS")]] <- extract_MDS(get(res))
}
# saveRDS(plots_MDS_PCA, "Figures/Plots_MDS_PCA_vsOthers_lncRNA.RDS")
# saveRDS(plots_MDS_PCA, "Figures/Plots_MDS_PCA_vsOthers.RDS")
# saveRDS(plots_MDS_PCA, "Figures/Plots_MDS_PCA_vsNBM.RDS")
```


```{r}
heatmaps <- list()
for (res in results){
 heatmaps[[res]] <- get(res)$Heatmap
}

# saveRDS(heatmaps, "Figures/Heatmaps_vsOthers_lncRNA.RDS")
# saveRDS(heatmaps, "Figures/Heatmaps_vsOthers.RDS")
# saveRDS(heatmaps, "Figures/Heatmaps_vsNBM.RDS")
```

plots were saved in the plotting_kallisto_RBD.Rmd script.

#Gene-Set Enrichment Analysis 

```{r}
library(rslurm)
```

```{r}
sopt <- list('nodes'='1', 
             'partition'='largenode',
             'cpus-per-task'='16',
             'ntasks'='1',
             'mem'='150G',
             'time' = '24:00:00',
             'mail-type'='FAIL,END',
             'mail-user'='jlsmith3@fredhutch.org') 

```

```{r}
# results <- ls(pattern = ".+vs.+") 
results <- dir(paste(getwd(),"RData",sep="/"), pattern = "Others_lncRNA.RDS|AMLvsNBM_lncRNA", #Others
               full.names = TRUE,
               include.dirs = TRUE, 
               recursive = TRUE) %>% 
  grep("GAGE",., value=TRUE, invert=TRUE)

results[1:2]
length(results)
```

```{r}
CAT <- readRDS("~/RNA_seq_Analysis/0000.00.01_GSEA_geneSets_gmt/FANTOM_CAT_v1_Disease_Cell_Ontologies.RDS")
length(CAT)
# head(CAT)
```

```{r warning=FALSE}
path="/fh/scratch/delete90/meshinchi_s/jlsmith3/kallisto"
setwd(path)

GAGE.Res <- list()
for(res in results){
  
  n <- str_split_fixed(res, pattern = "/", n=13)[,13] %>%
      gsub("\\.RDS", "", .)

  res <- readRDS(res)
  
  pheno <- res$phenovector
  mat <- rmDups(count.matrix = res$DE$Voom$E,
                ID.map = select(anno, gene_name,gene_id),
                rowname.GeneSym=FALSE)
  
  ref <- which(colnames(mat) %in%
                   names(pheno[pheno == "GroupB"]))

  cond <- which(colnames(mat) %in%
                    names(pheno[pheno == "GroupA"]))

  # r <- GAGE_GSA(df= mat, ref=ref,
  #               samp=cond,type="expn",geneset = CAT)
  # 
  # GAGE.Res[[n]] <- r
  # rm(res)
  
  #rSlurm was not working at all for some reason. just hanging. 
  slurm_call(f=GAGE_GSA,
            jobname = paste0(n, "_CAT_GSEA"),
            params =  list(df= mat,
                           ref=ref,
                           samp=cond,
                           type="expn",
                           geneset=CAT),
            add_objects = c("mat","CAT"),
            slurm_options=sopt,
            submit = TRUE) #Submitted batch job
  
}

# saveRDS(GAGE.Res,"TARGET_AML_FANTOM_CAT_GSEA_temp.RDS")
```


```{r}
gsea.res <- dir(path=path,
                recursive = TRUE,  
                pattern = "results_0.RDS", 
                include.dirs = TRUE) %>% 
  grep("_CAT_GSEA", ., value=TRUE) %>% 
  paste0(path, "/", .)

gsea.res
```

```{r}
for(i in 1:length(gsea.res)){
  name <- gsub("^.+\\/_rslurm_(.+)\\/.+.RDS", "\\1", gsea.res[i])
  print(name)
  assign(name, readRDS(gsea.res[i]))
}
```

```{r}
results <- ls(pattern="GSEA")
for (i in 1:length(results)){
  name1 <- paste0("GAGE/",results[i],"_Upregulated.csv")
  name2 <- paste0("GAGE/",results[i],"_Downregulated.csv")
  # print(c(name1, name2))
  
  # get(results[i])$SigPaths.Up[,1:5] %>%
  #   write.csv(., file=name1, row.names = TRUE)
  # 
  # get(results[i])$SigPaths.Dn[,1:5] %>%
  #   write.csv(., file=name2, row.names = TRUE)
}
```


results <- names(GAGE.Res)
for (i in 1:length(results)){
  name1 <- paste0("GAGE/", results[i],"_Upregulated_CAT.csv")
  name2 <- paste0("GAGE/",results[i],"_Downregulated_CAT.csv")
  # print(c(name1))
  
  GAGE.Res[[results[i]]]$SigPaths.Up[,1:5] %>%
    write.csv(x = ., file=name1, row.names = TRUE)

  GAGE.Res[[results[i]]]$SigPaths.Dn[,1:5] %>%
    write.csv(x=., file=name2, row.names = TRUE)
}




#Session Information 

```{r}
sessionInfo()
```

