---
title: 'IL3RA Expression in Fusion Groups in Ribodepleted RNAseq'
author: "Jenny Smith"
date: "April 5, 2019"
output: html_document
---



#Set-up 

```{r setup}
library(knitr)
knitr::opts_chunk$set(tidy.opts=list(width.cutoff=50),tidy=TRUE, fig.align='center', fig.height=5, fig.width=8, dpi = 600)
knitr::opts_knit$set(root.dir = '/fh/fast/meshinchi_s/workingDir/TARGET/AML_TARGET/RNA/mRNAseq/analysis/2019.04.05_Kallisto_DE_Transcriptome_MS/')
options(stringsAsFactors = FALSE)
```

```{r message = FALSE, warning=FALSE}
library(stringr)
library(magrittr)
library(ggplot2)
library(dplyr)
library(tibble)
library(tidyr)
# library(XLConnect)
library(DeGSEA, lib.loc = "/home/jlsmith3/R/x86_64-pc-linux-gnu-library/3.5") #my custom build R package that used to be imported through `source()` 
library(ggalluvial)
getwd()
```



#Define Functions 

```{r}
#Function to split the fusion columns into multiple single columns, coded as Yes/No/Unknown
createMultiCols <- function(col,split=FALSE,suffix){
  #col is the column with many factors (eg all fusions). character vector lenght 1.
  #suffix is a character string for the suffix on all column names
  #desinged for dplyr bind_cols() function
  
  #example
  # df <- df %>%
  #   bind_cols(.,createMultiCols(.$Fusion, suffix=".sensefusion")) 
  
  if(split){
    groups <- unique(gsub(" ","",unlist(str_split(col, "; "))))
    groups <- groups[groups != ""]
  }else{
    groups <- unique(col)
  }
  list <- lapply(groups, function(x) ifelse(grepl(paste0(x, ";"), col) | grepl(paste0(x, "$"), col) , "Yes",
                                            ifelse(grepl("^$|Unknown",col) | is.na(col), "Unknown","No"))) 
  list.names <- gsub("-", ".",  groups)
  names(list) <- paste0(list.names,suffix)
  return(list)
}

```


```{r}
classify_fusions <- function(FOI, ISCN){
  require(dplyr)
  require(tibble)
  
  ref <- data.frame(matrix(ncol = 3, nrow = 0,
                           dimnames = list(NULL,c("Fusion","Regex","Cyto.Regex")))) %>% 
    add_row(Fusion=c("NUP98-NSD1",
                    "NUP98-KDM5A",
                    "CBFA2T3-GLIS2",
                    "KMT2A-MLLT3", 
                    "KMT2A-MLLT10",
                    "KMT2A-MLLT4",
                    "KMT2A-ELL",
                    "KMT2A-MLLT1", 
                    "RUNX1-RUNX1T1", 
                    "CBFB-MYH11"), 
           Regex= c("NUP98-NSD1|NSD1-NUP98", 
                    "NUP98-KDM5A|KDM5A-NUP98", 
                    "CBFA2T3-GLIS2|GLIS2-CBFA2T3", 
                    "KMT2A-MLLT3|MLLT3-KMT2A", 
                    "KMT2A-MLLT10|MLLT10-KMT2A", 
                    "KMT2A-MLLT4|MLLT4-KMT2A", 
                    "KMT2A-ELL|ELL-KMT2A", 
                    "KMT2A-MLLT1(;|$)|MLLT1-KMT2A(;|$)", 
                    "RUNX1-RUNX1T1|RUNX1T1-RUNX1", 
                    "CBFB-MYH11|MYH11-CBFB"), 
           Cyto.Regex=c("t\\(\\d{0,2};?5;11;?\\d{0,2}\\)\\([pq]?.{0,};?p35;p15.{0,}\\)", 
                        "t\\(\\d{02};?11;12;?\\d{0,2}\\)\\([pq]?.{0,};?p15;p13.{0,}\\)", 
                        "inv\\(16\\)\\(p13;?q24\\)", 
                        
                        "t\\(\\d{0,2};?9;11;?\\d{0,2}\\)\\s?\\([pq]?.{0,};?p2\\d;q23.{0,}\\)", 
                        "(t|dic|ins)\\(\\d{0,2};?10;11;?\\d{0,2}\\)\\([pq]?.{0,};?p1\\d;q23.{0,}\\)",  #46,XY,ins(10;11)(p12;q23q13)[18]/46,XY[2]
                        "t\\(\\d{0,2};?6;11;?\\d{0,2}\\)\\([pq]?.{0,};?q2\\d;q23.{0,}\\)", 
                        
                        "t\\(\\d{0,2};?11;19;?\\d{0,2}\\)\\([pq]?.{0,};?q23;?p13.1.{0,}\\)", 
                        "t\\(\\d{0,2};?11;19;?\\d{0,2}\\)\\([pq]?.{0,};?q23;?p13.[23].{0,}\\)", 
                        "t\\(\\d{0,2};?8;21;?\\d{0,2}\\)\\(q22;?q22\\)|ins\\(21;8\\)|ins\\(8;21\\)", 
                        "inv\\(16\\)\\(p13.{0,3}q22\\)|t\\(16;16\\)\\(p13.{0,3}q22|ins\\(16;16\\)\\(p13.1;?q22")) %>% 
    set_rownames(.$Fusion)
  
  res <- NULL
  for(r in 1:nrow(ref)){
    df <- ref[r, ]
  
    #Data Query
    FOI.Present <- any(grepl(df$Regex,FOI))
    Cyto.Present <- grepl(df$Cyto.Regex, ISCN)
    No.Cyto <- is.na(ISCN)

    #define classifications
    if(FOI.Present & No.Cyto){
      res <- c(res,"RNA seq only; no cyto avail")
    }else if(FOI.Present & Cyto.Present){
      res <- c(res,"both confirms")
    } else if(FOI.Present & !Cyto.Present){
      res <- c(res,"RNA seq only")
    } else if(!FOI.Present & Cyto.Present){
      res <- c(res,"cyto only")
    }else if(!FOI.Present & ! Cyto.Present){
      res <- c(res,NA)
    }
  }

 names(res) <- ref[,"Fusion"]
 if(sum(!is.na(res)) == 0){
   res <- NA 
   names(res) <- "none"
 }else if(sum(!is.na(res)) == 1 ){
   res <- res[! is.na(res)]
   
 }else if(sum(!is.na(res)) > 1){
   n <- paste(names(res[! is.na(res)]),collapse = "; ")
   res <- "fusion/cyto data conflict"
   names(res) <- n
 }
  
  res <- paste(names(res), res, sep=": ")
  return(res)
}
```


```{r}
rmDups <- function(count.matrix, ID.map, matrix.class=TRUE){
  
  df <- count.matrix %>%
    as.data.frame() %>%
    rownames_to_column("gene_id") %>%
    left_join(., ID.map, by=c("gene_id")) %>% 
    select(gene_id, gene_name, everything()) %>%
    filter(!grepl("_PAR_", gene_id)) #specific to gencode refs
  
  
  dup <- df$gene_name[which(duplicated(df$gene_name))]
  ddf <- df %>%
    filter(gene_name %in% dup) %>% 
    arrange(gene_name)%>% 
    mutate(Variance=genefilter::rowVars(select(., -gene_id, -gene_name))) %>%

    group_by(gene_name) %>% 
    mutate(High.Variance=Variance==max(Variance)) %>% 
    ungroup() %>% 
    
    filter(High.Variance)  %>%
    filter(!duplicated(gene_name)) %>% #if variance tied, pick first one
    select(-gene_id, -Variance,-High.Variance, everything()) 
  
  rmDups.df <- df %>% 
    filter(! gene_name %in% dup) %>% 
    select(-gene_id) %>%
    bind_rows(.,ddf) %>%
    column_to_rownames("gene_name")
  
  if(matrix.class){
    rmDups.df <- data.matrix(rmDups.df)
  }
  
  return(rmDups.df)
}  
```


#Read in the Clinical Data

```{r}
merged <- read.csv("~/reference_mapping-files/TARGET_AML_0531_1031_merged_CDEs_3.01.19.csv")


merged <- merged %>% 
  filter(!is.na(USI)) %>% 
  set_rownames(.$USI)

# head(merged[,1:5])
dim(merged)
```

```{r}
manifest <- read.csv("/fh/fast/meshinchi_s/workingDir/TARGET/AML_TARGET/SequencingDataMatrix/TARGET_AML_0531_1031_miRNAseq_Ribodepletion_mRNAseq_Manifest_v4.csv")

# head(manifest)
```

```{r}
primary.fusion <- read.csv("~/RNA_seq_Analysis/2019.04.22_Transcriptome_Profile_AML/Final_list_AAML1031_fusions_with_cyto_information.csv")

# head(primary.fusion)
dim(primary.fusion)
```

```{r}
table(primary.fusion$Primary.Fusion.CNV) %>% 
  as.data.frame() %>% 
  arrange(desc(Freq)) %>% 
  head()
```

```{r}
all_fusions <- read.csv("/fh/fast/meshinchi_s/workingDir/TARGET/AML_TARGET/RNA/mRNAseq/level2/fusion/concat_matrices/2017July_BCCA_0531_1031_TransAbyss_STAR_K.Mungall_Combined_Illumina_data/TARGET_AML_0531_1031_TransAbyss_STAR_TargetedAlignment_Condensed_Fusion_Columns_4.15.19.csv") %>% 
  mutate(USI=str_split_fixed(Patient, "\\.", n=5)[,3])

# head(all_fusions)
```


#Define Fusion Groups on RNA-seq and Cytogenetic information

1. NUP98-NSD1 t(5;11)(q35;p15.5)
2. NUP98-KDM5A t(11;12)(p15;p13) 
3. CBFA2T3-GLIS2 inv(16)(p13q24) 
MLL  11q23
  4. KMT2A-MLLT3 t(9;11)(p21;q23)
  5. KMT2A-MLLT10 t(10;11)(p12;q23)
  6. KMT2A-MLLT4 t(6;11)(q27;q23) 
  7. KMT2A-ELL t(11;19)(q23;p13.1)
  8. KMT2A-MLLT1 t(11;19)(q23;p13.3) 
9. RUNX1-RUNX1T1 t(8;21)
10.CBFB-MYH11 inv(16)

```{r warning=FALSE}
fusions.info <- manifest %>% 
  
  #Select columns and clean 
  select(PATIENT_ID_Original, Final_Patient_ID, 
         USI, Reg.,Replicate,Type, Batch) %>% 
  filter(Type != "CellLine") %>%
  
  #Add replicate information 
  mutate(Replicate=ifelse(grepl("Rep", Replicate), "_replicate", "")) %>% 
  unite(col = "Colname",Final_Patient_ID, Replicate, sep = "", remove = FALSE ) %>% 
  filter(!grepl("PATGIG.+rep|PATISD.+rep", Colname)) %>% #duplicates 
  
  #Make Ptient barcodes the same
  mutate_all(funs(gsub("-", "\\.", .))) %>% 
  mutate(Reg.=as.character(Reg.)) %>%
  
  #Add in curated fusion info from ISCN and RNAseq by Rhonda
  left_join(.,select(primary.fusion,Reg.=Patient.registration.number,
                     Primary.Fusion.CNV, Cyto.vs..Seq,
                     Additional.Fusions.CNV ,Comments, 
                     Protocol=Study, ISCN) %>%
              mutate(Reg.=as.character(Reg.)) %>% #Study, ISCN,
              filter( ! is.na(Reg.)), #filter only remove Stella to avoid duplicates
            by=c("Reg.")) %>%
  mutate_at(vars(Primary.Fusion.CNV, Cyto.vs..Seq), funs(gsub(" $", "", .))) %>%
  mutate(Reg.=ifelse(is.na(Reg.), USI, Reg.)) %>%
  
  #Add in the RNA-seq FOI information
  left_join(., select(all_fusions, Patient, FOI=FOI), 
            by=c("Colname"="Patient")) %>%
  
  #Add in ISCN and Study 
  left_join(., select(merged, Reg., Protocol,ISCN, Known.Fusion.detected.by.any.method) %>% 
              mutate(Reg.=as.character(Reg.)),
            by="Reg.") %>% 
  mutate(ISCN=ifelse(is.na(ISCN.x), ISCN.y, ISCN.x)) %>% 
  mutate(ISCN=ifelse(grepl("Unknown", ISCN) | ISCN == "", NA,  ISCN)) %>%
  mutate(Protocol=ifelse(is.na(Protocol.x), Protocol.y, Protocol.x)) %>% 
  select(-ISCN.x, -ISCN.y, -Protocol.x, -Protocol.y) 


# head(fusions.info)
dim(fusions.info) #1593   16
```

```{r warning=FALSE}
fusionGroup.info <- fusions.info %>% 
  
  dplyr::rowwise() %>%
  mutate(Cyto.vs..Seq_JSmith=classify_fusions(FOI=FOI, ISCN=ISCN)) %>%
  ungroup() %>%
  
  #keep only groups of interest for DE analysis, all others considered negative in pheno vector.
  filter(!grepl("none",Cyto.vs..Seq_JSmith)) %>% 
  
  separate(Cyto.vs..Seq_JSmith,
           into = c("Primary.Fusion_JSmith","Cyto.vs..Seq_JSmith"), sep=": ") %>% 
  mutate(Different=ifelse(!is.na(Cyto.vs..Seq) & Cyto.vs..Seq != Cyto.vs..Seq_JSmith, 
                          "Different", "Same")) %>%
  mutate(Cyto.vs..Seq_forPlot=case_when(
    grepl("PARVSF|PAYIET", USI) ~ Cyto.vs..Seq_JSmith, #found the ISCN or found batch 2 fusions calls so different than Rhonda 
    Different == "Different" & Primary.Fusion_JSmith == Primary.Fusion.CNV ~ Cyto.vs..Seq, 
    Different == "Different" & Primary.Fusion_JSmith != Primary.Fusion.CNV ~ Cyto.vs..Seq_JSmith, #I did't have monosomy7 or del5q take precedence
    is.na(Primary.Fusion.CNV) ~  Cyto.vs..Seq_JSmith,
    Different == "Same" & Primary.Fusion_JSmith == Primary.Fusion.CNV ~  Cyto.vs..Seq_JSmith,
    TRUE ~ "WTH")) %>% 
  
  #Reorder and arrange columns 
  select(1:8, matches("Primary.Fusion|Cyto.vs"),Different, everything(), -Replicate) %>% 
  arrange(Batch,Primary.Fusion_JSmith,Cyto.vs..Seq_JSmith) %>% 
  
  #I will avoid making executive decision here. will update once Rhonda/Jason Farrar can decide. 
  filter(Cyto.vs..Seq_JSmith != "fusion/cyto data conflict") %>% 
  bind_cols(.,createMultiCols(.$Primary.Fusion_JSmith, suffix=".forDEGs"))


head(fusionGroup.info)
# dim(fusionGroup.info) #755  28
```

```{r}
# table(fusionGroup.info$Batch) #565 in batch 1 and 197  in batch 2
table(fusionGroup.info$Cyto.vs..Seq_JSmith,fusionGroup.info$Different)
table(fusionGroup.info$Primary.Fusion_JSmith, useNA = "always")
```

```{r}
# write.csv(fusionGroup.info, "TARGET_AML_RBD_Fusion_Cyto_vs_RNAseq_GroupsOfInterest_.csv", row.names = FALSE)
```



table(fusion.info$Cyto.vs..Seq, useNA = "always")
table(fusion.info$Primary.Fusion.CNV, useNA = "always")
table(fusion.info$RNASeq_FOI, useNA = "always")
table(filter(fusion.info, grepl("KMT2A", Primary.Fusion.CNV)) %>% select(Primary.Fusion.CNV), useNA = "always") %>% as.data.frame() %>% filter(Freq >= 3) %>% arrange(desc(Freq))

table(filter(fusion.info, grepl("NSD1", RNASeq_FOI)) %>% select(RNASeq_FOI), useNA = "always") %>% as.data.frame() #%>% filter(Freq >= 3) %>% arrange(desc(Freq))  
table(filter(fusion.info, grepl("KDM5A", RNASeq_FOI)) %>% select(RNASeq_FOI), useNA = "always") %>% as.data.frame() #%>% filter(Freq >= 3) %>% arrange(desc(Freq))  


##Alluvial Plot 





#Read in the Gene Annotation References

https://bioconductor.org/packages/release/bioc/vignettes/tximport/inst/doc/tximport.html#limma-voom

```{r eval=FALSE}
anno <- read.csv("~/RNA_seq_Analysis/0000.00.02_Reference_GeneInfo/gencode.v29_RepBase.v24.01_GeneLevel_IDmap_anno.csv") %>% 
  select(1:2)

head(anno)
dim(anno)
```

```{r}
lncRNA <- read.csv("~/RNA_seq_Analysis/0000.00.02_Reference_GeneInfo/gencode.v29.lncRNAs_Transcript.IDmap.csv") %>%
  select(gene_id, gene_type, gene_name) %>%
  mutate_at(vars(gene_id),
            funs(gsub("\\.[0-9]{1,}$", "", .))) %>%
  unique()

head(lncRNA)
dim(lncRNA) #16066  lncRNA genes   3

# write.csv(lncRNA, "~/RNA_seq_Analysis/0000.00.02_Reference_GeneInfo/gencode.v29.lncRNAs_GeneLevel.IDmap.csv")
```

```{r}
table(lncRNA$gene_type)
```


#Read in the counts

```{r}
gene_TPM <- readRDS("~/RNA_seq_Analysis/0000.00.03_Expression_Matrices/TARGET_AML_RBD_Kallisto_Quant_GeneLevel_TPM.RDS")

rownames(gene_TPM) <- gsub("\\.[0-9]{1,}$", "", rownames(gene_TPM))

head(gene_TPM[,1:5])
dim(gene_TPM)

temp <- gene_TPM["ENSG00000185291",]

quantile(temp)
```

The gene counts are fairly low and the 


```{r}
TPM <- readRDS("~/RNA_seq_Analysis/0000.00.03_Expression_Matrices/TARGET_AML_RBD_Kallisto_Quant_TranscriptLevel_TPM.RDS")

#remove version numbers from the counts
rownames(TPM) <- gsub("\\.[0-9]{1,}$", "", rownames(TPM))


head(TPM[,1:5])
dim(TPM)
```

```{r}
ID.map <- read.csv("~/RNA_seq_Analysis/0000.00.02_Reference_GeneInfo/gencode.v29_RepBase.v24.01_Gene_IDmap_anno.csv")

head(ID.map)
```

#Subset for IL3RA Expression

```{r}
IL3RA <- TPM %>% 
  rownames_to_column("Transcript") %>% 
  left_join(.,ID.map, by=c("Transcript"="transcript_id")) %>%
  filter(grepl("IL3RA", gene_name)) %>% 
  select(Transcript, gene_name, gene_id, matches("ensembl|tmhmm|Compart"), everything(),
         -transcript_name, -Has.ADC.Developed) %>% 
  gather(USI,TPM, BM3897:S) %>% 
  mutate(Group=ifelse(grepl("^BM|^RO", USI), "NBM", "AML"))
  
head(IL3RA)  
dim(IL3RA)
```

```{r}
IL3RA_dom <- IL3RA %>%

  group_by(USI) %>%
  mutate(DominantSpecies=ifelse(TPM==max(TPM), Transcript, "NotDominant")) %>%
  mutate(DominantSpecies=ifelse(TPM==max(TPM) & max(TPM) == 0,"NotExpressed",DominantSpecies)) %>%
  
  #messily fix ties. Pick the first one
  mutate(D2=case_when(
    sum(grepl("^ENST", DominantSpecies)) > 1 ~ DominantSpecies[grep("^ENST", DominantSpecies)[1]], 
    TRUE ~ DominantSpecies)) %>% 
  
  #change any entries with two "ENST" dominant species to "not dominant" status
  group_by(USI,DominantSpecies) %>%
  mutate(D3=case_when(
    sum(grepl("^ENST", D2)) == n() & DominantSpecies != unique(D2)  ~ "NotDominant",
    TRUE ~ DominantSpecies)) %>%
  ungroup() %>%
  
  #change the column Domincant species the intermediate column called D3
  mutate(DominantSpecies=D3) %>%
  dplyr::select(USI,Transcript,TPM,DominantSpecies,-D2,-D3) %>%

  #makes NAs to zeros and collapse the rows using sum()
  group_by(USI) %>%
  spread(Transcript, TPM) %>%
  mutate_at(vars(matches("^ENST")), funs(replace(.,is.na(.),0))) %>%
  mutate_at(vars(matches("^ENST")), funs(sum)) %>%
  ungroup() %>%

  #remove not dominant entries (these are duplicates at this point) and define groups
  filter(DominantSpecies != "NotDominant") %>%
  mutate(Group=case_when(
    grepl("BM[0-9]|RO[0-9]", USI) ~ "NBM",
    TRUE ~ "AML")) %>%

  #sum the transcripts TPMs to get a total  count for the gene level expression
  dplyr::select(USI,Group, everything()) %>%
  mutate(Total_IL3RA=dplyr::select(., contains("ENST")) %>% rowSums()) %>%

  # #Add annotations from the clincial data
  left_join(., select(manifest, USI, Reg.) %>% 
              unique(),
            by="USI")  %>%
  left_join(., dplyr::select(merged,
                      Reg.,ISCN,
                      FAB_or_WHO.Classification,
                      Primary.Cytogenetic.Code) %>% 
              mutate(Reg.=as.character(Reg.)),
            by=c("Reg.")) %>%
  mutate_at(vars(FAB_or_WHO.Classification,Primary.Cytogenetic.Code), 
            funs(case_when(
              Group == "NBM" ~ "NBM",
              . == "" | is.na(.) ~  "Unknown",
              TRUE ~ .))) %>%
  mutate(Primary.Cytogenetic.Code=factor(Primary.Cytogenetic.Code, 
                                         levels = c("NBM",  "Other",   "Normal",  "MLL","t(8;21)", "inv(16)", "Unknown"))) %>%
  mutate(Missing_Exons_Primarily=ifelse(DominantSpecies != "ENST00000331035", "MissingE3E4", "HasE3E4")) %>%

  dplyr::select(USI,Group,ISCN,
                FAB_or_WHO.Classification,Primary.Cytogenetic.Code,Missing_Exons_Primarily, everything())
  
  
options(scipen = 999)
head(IL3RA_dom)
# dim(IL3RA_dom) #1462   14
```

```{r}
# write.csv(IL3RA_dom, "TARGET_AML_RBD_Kallisto_GRCh38_IL3RA_TPM.csv", row.names = FALSE)
```

```{r}
table(IL3RA_dom$Primary.Cytogenetic.Code, useNA = "always")
table(IL3RA_dom$Missing_Exons_Primarily)
```

```{r}
bar.all <- ggplot(IL3RA_dom, aes(x=DominantSpecies)) +
  geom_bar(position = position_dodge(), fill="darkblue") +
  labs(title="IL3RA Dominant Transcript Species", y="Number of Patients") +
  theme_classic() +
  theme(text=element_text(size=15))

bar.all
# ggsave(plot=bar.all, filename = "TARGET_AML_RBD_IL3RA_Transcripts_barplot.png", device = "png", height = 5, width = 7, units="in", dpi=100)
```

```{r}
table(IL3RA_dom$DominantSpecies)
```

```{r fig.width=10}
bar.by.group <- ggplot(IL3RA_dom, aes(x=DominantSpecies, fill=Primary.Cytogenetic.Code)) +
  geom_bar(position = position_dodge()) +
  labs(title="IL3RA Dominant Transcript Species", y="Number of Patients") +
  theme_classic() +
  theme(text=element_text(size=15))


bar.by.group
ggsave(plot=bar.by.group, filename = "TARGET_AML_RBD_IL3RA_byCyto_barplot.png", device = "png", height = 5, width = 10, units="in", dpi=100)
```

```{r}
max(IL3RA$TPM)
```

```{r}
max(IL3RA_dom$Total_IL3RA)
```


#Boxplot

```{r}
bplot <- ggplot(data=IL3RA, 
       mapping=aes(x=Transcript, y=log2(TPM+1), fill=Transcript, color=Group)) +
       geom_boxplot() +
  theme_classic() + 
  theme(text = element_text(size=15), 
        axis.text.x = element_text(angle=30, vjust=1, hjust=1), 
        plot.margin = ) +
  scale_color_manual(values = c("darkgrey","black"))

bplot
```

```{r}
ggsave()
```



#Session Information 

```{r}
sessionInfo()
```

