---
title: 'Differential Expression of Fusion Groups in Ribodepleted RNAseq'
author: "Jenny Smith"
date: "April 5, 2019"
output: html_document
---



#Set-up 

```{r setup}
library(knitr)
knitr::opts_chunk$set(tidy.opts=list(width.cutoff=50),tidy=TRUE, fig.align='center', fig.height=5, fig.width=8, dpi = 600)
knitr::opts_knit$set(root.dir = file.path(PROJHOME, '2019.04.05_Kallisto_DE_Transcriptome_MS/'))
options(stringsAsFactors = FALSE)
```

```{r message = FALSE, warning=FALSE}
library(stringr)
library(magrittr)
library(ggplot2)
library(dplyr)
library(tibble)
library(tidyr)
library(DeGSEA)
# library(DeGSEA, lib.loc = "/home/jlsmith3/R/x86_64-pc-linux-gnu-library/3.5") #my custom build R package that used to be imported through `source()` 
library(ggalluvial)
library(readr)
getwd()
```



#Define Functions 

```{r}
rmDups <- function(count.matrix, ID.map, matrix.class=TRUE){
  #function to collapse duplicate genes using and ID map. Genes with identical symbols are selected by variance,
  #df is the gene expression matrix. genes as rownames, patients as columns 
  #IDmap is a data fram with gene_id for ensembl or other identifier and gene_name column for gene symbol. 
  
  df <- count.matrix %>%
    as.data.frame() %>%
    rownames_to_column("gene_id") %>%
    left_join(., ID.map, by=c("gene_id")) %>% 
    select(gene_id, gene_name, everything()) %>%
    filter(!grepl("_PAR_", gene_id)) #specific to gencode refs
  
  
  dup <- df$gene_name[which(duplicated(df$gene_name))]
  ddf <- df %>%
    filter(gene_name %in% dup) %>% 
    arrange(gene_name)%>% 
    mutate(Variance=genefilter::rowVars(select(., -gene_id, -gene_name))) %>%

    group_by(gene_name) %>% 
    mutate(High.Variance=Variance==max(Variance)) %>% 
    ungroup() %>% 
    
    filter(High.Variance)  %>%
    filter(!duplicated(gene_name)) %>% #if variance tied, pick first one
    select(-gene_id, -Variance,-High.Variance, everything()) 
  
  rmDups.df <- df %>% 
    filter(! gene_name %in% dup) %>% 
    select(-gene_id) %>%
    bind_rows(.,ddf) %>%
    column_to_rownames("gene_name")
  
  if(matrix.class){
    rmDups.df <- data.matrix(rmDups.df)
  }
  
  return(rmDups.df)
}  
```



#Read in the Clinical Data

```{r}
# merged <- read.csv("~/reference_mapping-files/00_Old/TARGET_AML_0531_1031_merged_CDEs_3.01.19.csv")
# merged <- read_csv(file.path(CDE,"Merged/TARGET_AML_0531_1031_merged_CDEs_9.4.19.csv"))
merged <-  read.csv(file.path(CDE,"Merged/TARGET_AML_0531_1031_merged_CDEs_7.08.20.csv"))

merged <- merged %>% 
  filter(!is.na(USI) , USI != "Unknown") %>% 
  set_rownames(.$USI)

head(merged[,1:5])
# dim(merged)
```

```{r}
manifest <- read.csv(file.path(TARGET, "SequencingDataMatrix/00_archive/TARGET_AML_RBD_0531_1031_miRNAseq_mRNAseq_Manifest_v5.csv"))

# head(manifest)
```

```{r}
DE.groups <- read.csv("TARGET_AML_RBD_Cohorts_forDE_5.20.19.csv")

dim(DE.groups)
```

```{r}
table(DE.groups$CBFA2T3.GLIS2.forDEGs)
```


#Expression Data 

```{r}
TPMs <- readRDS(file.path(HOME,"0000.00.03_Expression_Matrices/TARGET_AML_RBD_Kallisto_Quant_GeneLevel_TPM.RDS"))

rownames(TPMs) <- gsub("\\.[0-9]{1,}$", "", rownames(TPMs))
head(TPMs[,1:5])
dim(TPMs) 
```



#Read in the Gene Annotation References

https://bioconductor.org/packages/release/bioc/vignettes/tximport/inst/doc/tximport.html#limma-voom

```{r eval=FALSE}
anno <- read.csv(file.path(HOME, "0000.00.02_Reference_GeneInfo/gencode.v29_RepBase.v24.01_GeneLevel_IDmap_anno.csv")) 

head(anno)
dim(anno)
```

```{r}
lncRNA <- read.csv(file.path(HOME, "0000.00.02_Reference_GeneInfo/gencode.v29.lncRNAs_GeneLevel.IDmap.csv"),
                   row.names = 1)

head(lncRNA)
dim(lncRNA)
```

```{r}
# dir(file.path(PROJHOME,"0000.00.02_Reference_GeneInfo"))
ADC_CART <- read.csv(file.path(PROJHOME, "0000.00.02_Reference_GeneInfo/ADC_and_CARTcell_Targets_Database_ADCReview_clinicaltrialsGov.csv")) %>% 
  filter(!is.na(Gene.symbol.of.target..Final.)) %>% 
  group_by(Treatment.type, Gene.symbol.of.target..Final.) %>% 
  mutate_all(~paste(., collapse="; ")) %>% 
  ungroup() %>% 
  
  mutate_all(~gsub("NA;\\s|\\s(NA;){1,}|; \\sNA$", "", .)) %>% 
  unique() %>% 
  arrange(Gene.symbol.of.target..Final.) %>% 
  select(Treatment.type, Gene.symbol.of.target..Final., everything())


head(ADC_CART)
dim(ADC_CART) #178  18


# write.csv(ADC_CART,file.path(PROJHOME, "0000.00.02_Reference_GeneInfo/ADC_and_CARTcell_Targets_Database_ADCReview_rmDups_clinicaltrialsGov.csv"))
```


```{r}
# table(duplicated(ADC_CART$Gene.symbol.of.target..Final.)) #ADC or CART available
```




## FANTOM CAT Pathways

```{r}
cell.ont <- read.delim("Resources_References/nature21374-s2/supp_table_10.sample_ontology_information.tsv", 
                       header=T)

head(cell.ont)
dim(cell.ont) #347   9
```

```{r}
disease.ont <- read.delim("Resources_References/nature21374-s2/supp_table_12.trait_information.tsv",
                          header=T)

head(disease.ont)
dim(disease.ont) #817  11
```

```{r}
gs.cell <- str_split(cell.ont$associated_geneID, ",") %>% 
  lapply(., function(x) gsub("\\.[0-9]{1,2}$", "", x)) %>% 
  lapply(.,function(x) x[-grep("^CATG",x)]) %>%
  set_names(., gsub(" ", "_",cell.ont$sample_ontology_term))

# head(gs.cell)
# tail(gs)
length(gs.cell) #347
# saveRDS(gs.cell, "~/RNA_seq_Analysis/0000.00.01_GSEA_geneSets_gmt/FANTOM_CAT_v1_Cell_Ontologies.RDS")
```

```{r}
gs.dis <- str_split(disease.ont$associated_geneID, ",") %>% 
  lapply(., function(x) gsub("\\.[0-9]{1,}$", "", x)) %>% 
  lapply(.,function(x) x[-grep("^CATG",x)]) %>% 
  set_names(., gsub(" ", "_",disease.ont$trait_term))

length(gs.dis)
# saveRDS(gs.dis, "~/RNA_seq_Analysis/0000.00.01_GSEA_geneSets_gmt/FANTOM_CAT_v1_Disease_Ontologies.RDS")
```

```{r}
# CAT <- readRDS("~/RNA_seq_Analysis/0000.00.01_GSEA_geneSets_gmt/FANTOM_CAT_v1_Disease_Cell_Ontologies.RDS")
# length(CAT)
```



#Read in the DE Results

```{r}
files <- dir(path="RData", pattern = ".RDS", recursive=TRUE, full.names = TRUE) 

files
```

```{r}
for (i in 1:length(files)){
  
  name <- gsub("^.+\\/(.+).RDS", "\\1", files[i])
  print(name)
  assign(name, readRDS(files[i]))
  
}
```



##Extract DEGs 

```{r}
# Others_lncRNA <- dir(pattern = "Others_lncRNA.RDS", include.dirs = TRUE, recursive = TRUE)
# # Others_lncRNA
# 
# for(i in 1:length(Others_lncRNA)){
#   name <- gsub("^.+\\/(.+)\\.RDS", "\\1", Others_lncRNA[i])
#   print(name)
#   assign(name, readRDS(Others_lncRNA[i]))
# }

# AML.Others <- dir(pattern = "[A-Z0-9]vsOthers.RDS",
#                   include.dirs = TRUE, 
#                   recursive = TRUE)
# 
# for(i in 1:length(AML.Others)){
#   name <- gsub("^.+\\/(.+)\\.RDS", "\\1", AML.Others[i])
#   print(name)
#   assign(name, readRDS(AML.Others[i]))
# }

# AML.NBM <- dir(pattern = "vsNBM.RDS|AMLvsNBM_allGenes.RDS", include.dirs = TRUE, recursive = TRUE)
# # AML.NBM
# 
# for(i in 1:length(AML.NBM)){
#   name <- gsub("^.+\\/(.+)\\.RDS", "\\1", AML.NBM[i])
#   print(name)
#   assign(name, readRDS(AML.NBM[i]))
# }
```

Had java memory issues with XLConnect package... 

```{r eval=FALSE}
results <- ls(pattern = "*vs*")
for (res in results){
  print(res)
  
  if(grepl("vsOthers$", res)){
    df <- extract_DEGs(get(res), anno = TRUE, geneLevel = TRUE)
    # write.csv(df, paste0("DEGs/", res, "_DEGs.csv"), row.names=FALSE)

    # createSheet(others, name = res )
    # writeWorksheet(others, df, sheet = res)
  }else if (grepl("vsNBM$|AMLvsNBM_allGenes", res)){
    df <- extract_DEGs(get(res), anno = TRUE, geneLevel = TRUE)
    # write.csv(df, paste0("DEGs/", res, "_DEGs.csv"), row.names=FALSE)

    # createSheet(NBM, name = res)
    # writeWorksheet(NBM, extract_DEGs(get(res), anno = TRUE, geneLevel = TRUE), sheet = res)
  }else if(grepl("lncRNA", res)){
    
    df <- extract_DEGs(get(res)) %>%
      left_join(.,  lncRNA, by=c("gene"="gene_id"))
    
    # write.csv(df, paste0("DEGs/", res, "_DEGs.csv"), row.names=FALSE)

    # createSheet(lnc, name = res)
    # writeWorksheet(lnc, df, sheet = res)
  }
}

```

#Examine GSEA Pathways
```{r}
path="/fh/scratch/delete90/meshinchi_s/jlsmith3/kallisto"
gsea.res <- dir(path=path,
                recursive = TRUE,  
                pattern = "results_0.RDS", 
                include.dirs = TRUE) %>% 
  grep("_CAT_GSEA", ., value=TRUE) %>% 
  paste0(path, "/", .)

gsea.res
```

```{r}
for(i in c(1:length(gsea.res))[10]) {
  name <- gsub("^.+\\/_rslurm_(.+)\\/.+.RDS", "\\1", gsea.res[i])
  print(name)
  assign(name, readRDS(gsea.res[i]))
}
```

```{r}
get_lncRNAs <- function(GAGE.res,pathway,ID.map,DEGs=NULL){
  
 core.genes <- GAGE.res$essSets.Up$coreGeneSets[pathway]
 
 lncs <-  lapply(names(core.genes), function(x) filter(ID.map,  gene_id %in% core.genes[[x]]) %>% 
              mutate(Pathway=x)) %>%
  bind_rows()
 
 if(!is.null(DEGs)){
  lncs <-  lncs %>% 
    left_join(., rownames_to_column(DEGs, "gene_id") %>% 
              select(logFC, adj.P.Val,gene_id),
            by="gene_id")
 }
  
  return(lncs)
  
}
```

```{r}
NSD1.lncs <- get_lncRNAs(GAGE.res = NUP98NSD1vsOthers_lncRNA_CAT_GSEA, pathway = NUP98NSD1vsOthers_lncRNA_CAT_GSEA$essSets.Up$essentialSets, 
                         ID.map = lncRNA)

NSD1.lncs
```


```{r}
# AMLvsNBM_lncRNA_CAT_GSEA$essSets.Up$essentialSets
nonred <- AMLvsNBM_lncRNA_CAT_GSEA$essSets.Up$coreGeneSets[rownames(AMLvsNBM_lncRNA_CAT_GSEA$SigPaths.Up)]
```

```{r}
t <- lapply(names(nonred), function(x) filter(lncRNA,  gene_id %in% nonred[[x]]) %>% 
              mutate(Pathway=x)) %>%
  bind_rows() %>% 
  left_join(., rownames_to_column(AMLvsNBM$DE$DE, "gene_id") %>% 
              select(logFC, adj.P.Val,gene_id),
            by="gene_id")

dim(t)
head(t)

# write.csv(t, "TARGET_AML_RBD_AMLvsNBM_lncRNAs_in_Pathways.csv", row.names = FALSE)
```




#lncRNA Genomic Locations annotations

## Map Novel CAT lncRNAs to gene IDs

```{r message=FALSE}
library(plyranges)
library(GenomicFeatures)
library(rtracklayer)
```

```{r}
lncRNAs.db <- makeTxDbFromGFF(file = "/fh/fast/meshinchi_s/workingDir/TARGET/Reference_Data/GRCh38/gtf/gencode.v29.long_noncoding_RNAs.gtf")

lncRNAs.db
```

```{r}
lncRNAs.37 <- read.delim("/fh/scratch/delete90/meshinchi_s/jlsmith3/lncipedia_5_2_hg19.gff", sep="\t", comment.char = "#", header = FALSE) %>% 
  filter(V3 != "exon")

temp <- str_split(pull(lncRNAs.37,V9),pattern = ";")

head(temp)

# head(lncRNAs.37)
# 
# lncRNAs.37.db <- makeTxDbFromGFF(file="/fh/scratch/delete90/meshinchi_s/jlsmith3/lncipedia_5_2_hg19.gff")
# 
# lncRNAs.37.db
```

```{r}
genes.lnc <- genes(lncRNAs.db)
genes.lnc$gene_id2 <- gsub("\\.[0-9]{1,2}", "", genes.lnc$gene_id)
names(genes.lnc) <-  genes.lnc$gene_id2

genes.37.lnc <- genes(lncRNAs.db)
genes.37.lnc$gene_id2 <- gsub("\\.[0-9]{1,2}", "", genes.37.lnc$gene_id)
names(genes.37.lnc) <- genes.37.lnc$gene_id2

head(genes.37.lnc)
```

```{r}
CAT.db <- read.delim("/fh/scratch/delete90/meshinchi_s/jlsmith3/5BFANTOMCAT5DRobustgene.gff", header = FALSE) %>% 
  separate(V9, into=c("location", "class", "class2", "gene_id","Ref"), remove = T, sep="[\\|;]") %>% 
  mutate(gene_id2=gsub("\\.[0-9]{1,2}", "", gene_id)) %>% 
  filter(class2 != "coding_mRNA", grepl("CATG" ,gene_id))

CAT.gr <- GRanges(seqnames = CAT.db$V1,
                  ranges = IRanges(start=CAT.db$V4,end=CAT.db$V5), 
                  strand = CAT.db$V7, 
                  mcols=dplyr::select(CAT.db,gene_id, gene_id2, class,class2))
names(CAT.gr) <- CAT.db$gene_id2

CAT.gr
```

```{r}
mapped.CAT <- subsetByOverlaps(x = genes.37.lnc, ranges = CAT.gr, minoverlap=50L, type='equal')

mapped.CAT
```

```{r}
eRNAs_intersect <- join_overlap_intersect(x=genes.lnc, y=enhancers)
mcols(eRNAs_intersect)$width <- width(eRNAs_intersect)
```

```{r}
hist(mcols(eRNAs_intersect)$width, breaks = seq (0,3000, by = 50))
median(mcols(eRNAs_intersect)$width)
```

https://support.bioconductor.org/p/72656/

```{r}
hits <- findOverlaps(refGR, testGR)
overlaps <- pintersect(refGR[queryHits(hits)], testGR[subjectHits(hits)])
percentOverlap <- width(overlaps) / width(testGR[subjectHits(hits)])

quantile(percentOverlap*100)
```

```{r}
hits <- hits[percentOverlap > 0.28] #of 6,858 enhancer regions in w some overlaping lncRNA transcription, only 50% have > 28% of basepairs overlapping. so most have just under 1/3 covered by enahncer region
```


#Investigate Suvival Associated lncRNAs

```{r}
library(goseq)
library(GenomicFeatures)
```

```{r}
lncRNAs.db <- makeTxDbFromGFF(file = file.path(GENREFS,"GRCh38/gtf/gencode.v29.long_noncoding_RNAs.gtf"))
# lncRNAs.db

tx_by_lnc <- transcriptsBy(lncRNAs.db, by="gene")
names(tx_by_lnc) <-  gsub("\\.[0-9]{1,2}", "", names(tx_by_lnc))
gene.lens <- median(width(tx_by_lnc))

length(gene.lens)
head(gene.lens)
# quantile(gene.lens)
```

```{r}
gs.cell <- readRDS(file.path(HOME,"0000.00.01_GSEA_geneSets_gmt/FANTOM_CAT_v1_Cell_Ontologies.RDS"))
gs.dis <- readRDS(file.path(HOME,"0000.00.01_GSEA_geneSets_gmt/FANTOM_CAT_v1_Disease_Ontologies.RDS"))
```

```{r}
get_lncRNAs2 <- function(pathways,ID.map,DEGs=NULL){
  
 
 lncs <-  lapply(names(pathways),
                 function(x) filter(ID.map,  gene_id %in% pathways[[x]]) %>% 
                              mutate(Pathway=x)) %>%
  bind_rows()
 
 if(!is.null(DEGs)){
  lncs <-  lncs %>% 
    left_join(., rownames_to_column(DEGs, "gene_id") %>% 
              select(logFC, adj.P.Val,gene_id),
            by="gene_id")
 }
  
  return(lncs)
  
}
```

```{r}
lncRNA.universe <- read.csv("Survival/lncRNAs_Used_InModeling.csv") %>% #expressed lncRNAs
  filter(., !grepl("PAR_", x))

# gene.universe <- read_csv("Survival/TARGET_AML_RBD_TMMCPM_forModeling.csv")[[1]] #All Expressed Genes 
# gene.universe <- grep("^ENSG",gene.universe, value=TRUE) %>% 
#   grep("_PAR_Y",., invert = T, value=TRUE)
# length(gene.universe) # 25643


lncRNA.signature  <- read.csv("Survival/TARGET_AML_RBD_lncRNAs_UV_upreg_pvalue_LT0.1_Coef_7.17.19.csv") %>% 
  dplyr::select(gene_id=gene, everything())

head(lncRNA.signature)
```

```{r}
lnc.Ontologies <- get_lncRNAs2(pathways = c(gs.cell, gs.dis),
                               ID.map = lncRNA.signature) %>% 
  group_by(Pathway) %>%
  mutate(Number.In.Pathway=dplyr::n()) %>%
  ungroup() %>%
  arrange(desc(Number.In.Pathway))


dim(lnc.Ontologies) #308   8
head(lnc.Ontologies)
# write.csv(lnc.Ontologies, "Survival/TARGET_AML_RBD_lncRNAs_Coef_Pathways_7.25.19.csv")
```

```{r}
lncSig <- ifelse(names(gene.lens) %in% lncRNA.signature$gene_id, 1,0) %>%
  set_names(names(gene.lens))

head(lncSig)
table(lncSig)
```

```{r}
CAT.cell <-  lapply(names(gs.cell), function(x) data.frame(gene_id=gs.cell[[x]],
                                                     Pathways=rep(x, length(gs.cell[[x]])))) %>% 
  bind_rows() %>% 
  filter(! grepl("^CATG", gene_id)) #%>%

# head(CAT.cell)
# dim(CAT.cell)
length(unique(CAT.cell$Pathways)) #347



CAT.dis <-  lapply(names(gs.dis), function(x) data.frame(gene_id=gs.dis[[x]],
                                                     Pathways=rep(x, length(gs.dis[[x]])))) %>% 
  bind_rows() %>% 
  filter(! grepl("^CATG", gene_id)) #%>%

# head(CAT.dis)
# dim(CAT.dis)
length(unique(CAT.dis$Pathways)) #347
```

```{r}
pwf <- nullp(DEgenes = lncSig, 
              genome = "hg19", 
              id = "ensGene", 
              bias.data = gene.lens,
              plot=FALSE)

GO.cell <- goseq(pwf, 
                 genome = "hg19",
                 id="ensGene", 
             gene2cat = CAT.cell, 
             use_genes_without_cat=TRUE)

GO.cell.filt <-  GO.cell %>% 
  filter(numInCat < 500 & numInCat >= 20) %>%
  arrange(over_represented_pvalue) %>%
  mutate(FDR_Over=p.adjust(over_represented_pvalue, method = "BH"),
         FDR_Under=p.adjust(under_represented_pvalue, method = "BH")) %>%
  filter(FDR_Over < 0.25 | FDR_Under < 0.25)


head(GO.cell.filt,n=20)
# tail(GO.cell.filt)
dim(GO.cell.filt) #13
```

```{r}
GO.cell %>% 
  filter(grepl("hema", category))
```

```{r}
GO.dis <- goseq(pwf, 
                 genome = "hg19",
                 id="ensGene", 
             gene2cat = CAT.dis, 
             use_genes_without_cat=TRUE)

GO.dis.filt <-  GO.dis %>% 
  filter(numInCat < 500 & numInCat >= 20) %>%
  arrange(over_represented_pvalue) %>%
  mutate(FDR_Over=p.adjust(over_represented_pvalue, method = "BH"),
         FDR_Under=p.adjust(under_represented_pvalue, method = "BH")) #%>%
  # filter(FDR_Over < 0.25 | FDR_Under < 0.25)


head(GO.dis.filt,n=20)
# tail(GO.cell.filt)
dim(GO.dis.filt) #13
```










#Examine NUP98-NSD1 

##Extract the NUP98-NSD1- like 

```{r fig.width=16}
plot(NUP98NSD1vsOthers_lncRNA$dendrogram$samp.c1)
```


```{r}
d.groups <- colorDends_Groups(dendrogram = NUP98NSD1vsOthers_lncRNA$dendrogram,
                              phenovector = NUP98NSD1vsOthers_lncRNA$phenovector, k= 7,
                              branchCol = c(1,2,3,4,5,6,7),
                              colorcodes = c("GroupA"="red","GroupB"="darkgrey"))
```


```{r fig.width=16}
# png(filename = "NUP98-NSD1_DE_lncRNA_Dend_k7.png", height = 4, width = 16, res=100, units="in")
par(cex=0.25, cex.axis=5, cex.main=5, mar =  c(5, 4, 8, 2) + 0.1)
plot(d.groups$dend, main="Dendrogram of NUP98-NSD1 \n DE lncRNAs")
# dev.off()
# par(cex.axis = 10)
# plot(d.groups$split_dends[[1]])

```

```{r}
groups.df <- data.frame(USI=unlist(d.groups$group_labels)) %>% 
  mutate(Group=case_when(
    USI %in% d.groups$group_labels[[1]] ~ 1,
    USI %in% d.groups$group_labels[[2]] ~ 2,
    USI %in% d.groups$group_labels[[3]] ~ 3,
    USI %in% d.groups$group_labels[[4]] ~ 4, 
    USI %in% d.groups$group_labels[[5]] ~ 5,
    USI %in% d.groups$group_labels[[6]] ~ 6,
    USI %in% d.groups$group_labels[[7]] ~ 7)) %>% 
  mutate(Group_color=case_when(
    Group == 1 ~ "black", 
    Group == 2 ~ "red",
    Group == 3 ~ "green",
    Group == 4 ~ "blue", 
    Group == 5 ~ "cyan",
    Group == 6 ~ "magenta",
    Group == 7 ~ "yellow")) %>% 
  mutate(NUP98.NSD1=case_when(
    USI %in% names(NUP98NSD1vsOthers_lncRNA$phenovector[NUP98NSD1vsOthers_lncRNA$phenovector =="GroupA"]) ~ "NUP98-NSD1", 
   ! USI %in% names(NUP98NSD1vsOthers_lncRNA$phenovector[NUP98NSD1vsOthers_lncRNA$phenovector =="GroupA"]) & Group == 1 ~ "NUP98-NSD1-like",
   ! USI %in% names(NUP98NSD1vsOthers_lncRNA$phenovector[NUP98NSD1vsOthers_lncRNA$phenovector =="GroupA"]) & Group == 2 ~ "NUP98-NSD1-like",
   TRUE ~ "OtherAML"))


table(groups.df$Group)
table(groups.df$NUP98.NSD1)
# write.csv(groups.df,"TARGET_AML_RBD_NUP98-NSD1_like_Groups_k7.csv", row.names = FALSE)
```


##Look at Enhancer RNA 

```{r message=FALSE}
library(plyranges)
library(GenomicFeatures)
library(rtracklayer)
```

```{r}
lncRNAs.db <- makeTxDbFromGFF(file = file.path(GENREFS,"/GRCh38/gtf/gencode.v29.long_noncoding_RNAs.gtf"))

lncRNAs.db
```

```{r}
enhancers <- import("/fh/scratch/delete90/meshinchi_s/jlsmith3/lncRNA/F5.hg38.enhancers.bed")

enhancers
```

```{r}
genes.lnc <- genes(lncRNAs.db)
strand(genes.lnc) <- "*"
```

```{r}
eRNAs.all <- subsetByOverlaps(x = genes.lnc, ranges = enhancers, minoverlap=200L, type='any')
```

https://support.bioconductor.org/p/72656/

```{r}
hits <- findOverlaps(refGR, testGR)
overlaps <- pintersect(refGR[queryHits(hits)], testGR[subjectHits(hits)])
percentOverlap <- width(overlaps) / width(testGR[subjectHits(hits)])

quantile(percentOverlap*100)
```


```{r}
hits <- hits[percentOverlap > 0.28] #of 6,858 enhancer regions in w some overlaping lncRNA transcription, only 50% have > 28% of basepairs overlapping. so most have just under 1/3 covered by enahncer region
```

```{r}
eRNAs_intersect <- join_overlap_intersect(x=genes.lnc, y=enhancers)
mcols(eRNAs_intersect)$width <- width(eRNAs_intersect)
```

```{r}
hist(mcols(eRNAs_intersect)$width, breaks = seq (0,3000, by = 50))
median(mcols(eRNAs_intersect)$width)
```





##GSEA with lncRNA gene-sets 


```{r}
dim(NUP98NSD1vsOthers_lncRNA$DE$Voom$E)
head(NUP98NSD1vsOthers_lncRNA$DE$Voom$E[,1:5])
```



```{r}
GSA.lncRNA <- gage_from_pipeline(twoGroups_DEGs.res = NUP98NSD1vsOthers_lncRNA, 
                                 type="expn",
                                 geneset = c(gs.cell,gs.dis))

```

```{r}
summary(GSA.lncRNA)
```

```{r}
head(GSA.lncRNA$SigPaths.Up[,1:5])
```

```{r}
# saveRDS(GSA.lncRNA, "NUP98_NSD1_lncRNA_GAGE.RDS")


GSA.lncRNA <- readRDS("NUP98_NSD1_lncRNA_GAGE.RDS")
```

```{r}
# write.csv(GSA.lncRNA$SigPaths.Up,"TARGET_AML_RBD_NUP98-NSD1_CAT_UPReg_GSEA.csv")
# write.csv(GSA.lncRNA$SigPaths.Dn,"TARGET_AML_RBD_NUP98-NSD1_CAT_DNReg_GSEA.csv")
```



#NUP98-KDM5A

```{r}
KDM5A.DEGs <- read.csv("DEGs/NUP98KDM5AvsOthers_DEGs.csv") %>% 
  filter(logFC > 1)

# head(KDM5A.DEGs)
dim(KDM5A.DEGs) # 2189   15
```

```{r}
KDM5AvsNBM.DEGs <- read.csv("DEGs/NUP98KDM5AvsNBM_DEGs.csv") %>% 
  filter(logFC > 1)

dim(KDM5AvsNBM.DEGs) # 2394   15
```


##Annotate NUP98-KDM5A with ADCs 

```{r}
library(rDGIdb)
```

```{r}
ADCs <- read.csv(file.path(PROJHOME,"0000.00.02_Reference_GeneInfo",
                           "Full_ADC_Target_Database_rmDuplicates_08.28.2019.csv"))

# head(ADCs)
dim(ADCs)
```

```{r}
genes <- c(KDM5A.DEGs$gene_name, KDM5AvsNBM.DEGs$gene_name) %>% 
  unique() #3600 upregulated genes


DGI_Filter <- queryDGIdb(genes = genes,
                             geneCategories = c("CLINICALLY ACTIONABLE"))
  
DGI_Final <- detailedResults(DGI_Filter) %>% 
                  group_by(Gene) %>% 
                  #collapse genes with multiple drugs into a single row
                  mutate_at(vars(Drug:PMIDs),
                            ~collapseRows(col = ., uniq = FALSE, sep="; ")) %>% 
                  ungroup()  %>%
                  dplyr::select(-SearchTerm) %>%
                  unique() 
  


dim(DGI_Final) #18 hits 
```

```{r}
results <- select(KDM5A.DEGs, gene:logFC,adj.P.Val) %>% 
      rename_at(vars(logFC,adj.P.Val),~paste0(.,"_vsOtherAML")) %>% 
      full_join(.,select(KDM5AvsNBM.DEGs,gene,logFC,adj.P.Val) %>% 
                  rename_at(vars(logFC,adj.P.Val), ~paste0(.,"_vsNBM")), 
                by=c("gene"="gene")) %>%

      left_join(., ADCs,
                by=c("gene_name"="Gene.symbol.of.ADC.GRCh37.69")) %>%
      left_join(.,DGI_Final,
                by=c("gene_name"="Gene")) %>%
      dplyr::select(gene:adj.P.Val_vsNBM,
                    Antibody.Drug.Conjugate..ADC..symbol.or.trade.name,
                    Drug.Gene.Interaction.DB=Drug,
                    everything(), 
                    -Gene.symbol.of.ADC.target..Original.,-InteractionType) %>%
      rename_at(vars(Source:PMIDs), ~paste0(.,"_DGI.DB")) %>% 
      rename_at(vars(Drug.payload.name:Comments), ~ paste0(.,"_ADC")) %>%
        
      filter(!is.na(Antibody.Drug.Conjugate..ADC..symbol.or.trade.name) |
               !is.na(Drug.Gene.Interaction.DB))

head(results)
dim(results) #28 18

# write.csv(results,"TARGET_AML_NUP98.KDM5A_DEGs_withADCs_SmallMoleculeInhibitors.csv", row.names = FALSE)
```

```{r}
ADC.expn <- TPMs %>% 
  as.data.frame() %>%
  rownames_to_column("gene") %>% 
  filter(gene %in% results$gene) %>% 
  gather(USI,TPM,-gene) %>%
  left_join(.,select(results,gene, gene_name), by="gene") %>%
  mutate(Group=case_when(
    USI %in% filter(merged, NUP98.KDM5A == "Yes")$USI ~ "Positive", 
    grepl("^BM|^RO", USI) ~ "NBM",
    TRUE ~ "Negative"), 
    log2_TPM=log2(TPM+1)) %>% #log2_TPM=log2(TPM+1)
  arrange(gene,Group,desc(TPM)) %>%
  mutate(Order=row_number()) %>%
  ungroup()


dim(ADC.expn)
# length(unique(ADC.expn$gene)) #2,830 genes
head(ADC.expn[,])
```

```{r}
table(filter(ADC.expn, gene_name=="CD22")$Group)
```

```{r fig.width=14, fig.height=12}
ADC.boxplots <- ggplot(ADC.expn, aes(x=Group,y=log2_TPM, fill=Group)) + #color=Group
  geom_jitter(aes(color=Group)) +
  geom_boxplot(alpha=0.3, outlier.shape = NA) +
  # geom_bar(stat="identity", show.legend = TRUE) +
  facet_wrap(~gene_name, scales="free_y") +
  labs(x="Fusion Status") +
  theme_classic() +
  theme(axis.text.x = element_blank()) +
  scale_fill_manual(values = c("Positive"="firebrick","Negative"="navy", "NBM"="black")) +
  scale_color_manual(values = c("Positive"="firebrick","Negative"="dodgerblue1", "NBM"="grey40"))

ADC.boxplots
# ggsave("TARGET_AML_NUP98-KDM5A_ADC_Targets_boxplots.pdf",ADC.boxplots,device = "pdf", height = 12, width = 14,units = "in")
```

```{r fig.width=16, fig.height=12}
ADC.waterfall <- ggplot(ADC.expn, aes(x=Order,y=TPM, fill=Group,color=Group)) + 
  geom_bar(stat="identity", show.legend = TRUE) +
  facet_wrap(~gene_name, scales="free") +
  labs(x="Patient") +
  theme_classic() +
  theme(axis.text.x = element_blank()) +
  scale_fill_manual(values = c("Positive"="firebrick","Negative"="dodgerblue1", "NBM"="black")) +
  scale_color_manual(values = c("Positive"="firebrick","Negative"="dodgerblue1", "NBM"="grey40"))


ADC.waterfall
# ggsave("TARGET_AML_NUP98-KDM5A_ADC_Targets_Waterfalls.pdf",ADC.waterfall,device = "pdf", height = 12, width = 16,units = "in")
```


#KMT2A-MLLT1

```{r}
MLLT1.DEGs <- read.csv("DEGs/KMT2AMLLT1vsOthers_DEGs.csv") %>% 
  filter(logFC > 1)

dim(MLLT1.DEGs) #  1849   15
```

```{r}
MLLT1vsNBM.DEGs <- read.csv("DEGs/KMT2AMLLT1vsNBM_DEGs.csv") %>% 
  filter(logFC > 1)

dim(MLLT1vsNBM.DEGs) #2971  15
```


##Annotate KMT2A-MLLT1 with ADCs 

```{r}
library(rDGIdb)
```

```{r}
ADCs <- read.csv(file.path(PROJHOME,"0000.00.02_Reference_GeneInfo",
                           "Full_ADC_Target_Database_rmDuplicates_08.28.2019.csv"))

# head(ADCs)
dim(ADCs)
```

```{r}
genes <- c(MLLT1.DEGs$gene_name, MLLT1vsNBM.DEGs$gene_name) %>% 
  unique() #3616 upregulated genes


DGI_Filter <- queryDGIdb(genes = genes,
                             geneCategories = c("CLINICALLY ACTIONABLE"))
  
DGI_Final <- detailedResults(DGI_Filter) %>% 
                  group_by(Gene) %>% 
                  #collapse genes with multiple drugs into a single row
                  mutate_at(vars(Drug:PMIDs),
                            ~collapseRows(col = ., uniq = FALSE, sep="; ")) %>% 
                  ungroup()  %>%
                  dplyr::select(-SearchTerm) %>%
                  unique() 
  


dim(DGI_Final) #21 hits 
```

```{r}
results <- select(MLLT1.DEGs, gene:logFC,adj.P.Val) %>% 
      rename_at(vars(logFC,adj.P.Val),~paste0(.,"_vsOtherAML")) %>% 
      full_join(.,select(MLLT1vsNBM.DEGs,gene,logFC,adj.P.Val) %>% 
                  rename_at(vars(logFC,adj.P.Val), ~paste0(.,"_vsNBM")), 
                by=c("gene"="gene")) %>%

      left_join(., ADCs,
                by=c("gene_name"="Gene.symbol.of.ADC.GRCh37.69")) %>%
      left_join(.,DGI_Final,
                by=c("gene_name"="Gene")) %>%
      dplyr::select(gene:adj.P.Val_vsNBM,
                    Antibody.Drug.Conjugate..ADC..symbol.or.trade.name,
                    Drug.Gene.Interaction.DB=Drug,
                    everything(), 
                    -Gene.symbol.of.ADC.target..Original.,-InteractionType) %>%
      rename_at(vars(Source:PMIDs), ~paste0(.,"_DGI.DB")) %>% 
      rename_at(vars(Drug.payload.name:Comments), ~ paste0(.,"_ADC")) %>%
        
      filter(!is.na(Antibody.Drug.Conjugate..ADC..symbol.or.trade.name) |
               !is.na(Drug.Gene.Interaction.DB))

head(results)
dim(results) #20 18

# write.csv(results,"TARGET_AML_KMT2A.MLLT1_DEGs_withADCs_SmallMoleculeInhibitors.csv", row.names = FALSE)
```

```{r}
ADC.expn <- TPMs %>% 
  as.data.frame() %>%
  rownames_to_column("gene") %>% 
  filter(gene %in% results$gene) %>% 
  gather(USI,TPM,-gene) %>%
  left_join(.,select(results,gene, gene_name), by="gene") %>%
  mutate(Group=case_when(
    USI %in% filter(merged,  grepl("MLLT1",Primary.Fusion.CNV))$USI ~ "Positive", 
    USI %in% filter(merged,grepl("MLLT1", Additional.Fusions.CNV))$USI ~ "Positive",
    grepl("^BM|^RO", USI) ~ "NBM",
    TRUE ~ "Negative"), 
    log2_TPM=log2(TPM+1)) %>% #log2_TPM=log2(TPM+1)
  arrange(gene,Group,desc(TPM)) %>%
  mutate(Order=row_number()) %>%
  ungroup()


dim(ADC.expn)
length(unique(ADC.expn$gene)) #20 genes
head(ADC.expn[,])
```

```{r}
table(filter(ADC.expn, gene_name=="ETV1")$Group)
```

```{r fig.width=14, fig.height=12}
ADC.boxplots <- ggplot(ADC.expn, aes(x=Group,y=log2_TPM, fill=Group)) + #color=Group
  geom_jitter(aes(color=Group)) +
  geom_boxplot(alpha=0.3, outlier.shape = NA) +
  # geom_bar(stat="identity", show.legend = TRUE) +
  facet_wrap(~gene_name, scales="free_y") +
  labs(x="Fusion Status") +
  theme_classic() +
  theme(axis.text.x = element_blank()) +
  scale_fill_manual(values = c("Positive"="firebrick","Negative"="navy", "NBM"="black")) +
  scale_color_manual(values = c("Positive"="firebrick","Negative"="dodgerblue1", "NBM"="grey40"))

ADC.boxplots
# ggsave("TARGET_AML_KMT2A-MLLT1_ADC_Targets_boxplots.pdf",ADC.boxplots,device = "pdf", height = 12, width = 14,units = "in")
```

```{r fig.width=16, fig.height=12}
ADC.waterfall <- ggplot(ADC.expn, aes(x=Order,y=TPM, fill=Group,color=Group)) + 
  geom_bar(stat="identity", show.legend = TRUE) +
  facet_wrap(~gene_name, scales="free") +
  labs(x="Patient") +
  theme_classic() +
  theme(axis.text.x = element_blank()) +
  scale_fill_manual(values = c("Positive"="firebrick","Negative"="dodgerblue1", "NBM"="black")) +
  scale_color_manual(values = c("Positive"="firebrick","Negative"="dodgerblue1", "NBM"="grey40"))


ADC.waterfall
# ggsave("TARGET_AML_KMT2A-MLLT1_ADC_Targets_Waterfalls.pdf",ADC.waterfall,device = "pdf", height = 12, width = 16,units = "in")
```


#CBFA2T3-GLIS2


```{r}
table(duplicated(ADC_CART$Gene.symbol.of.target..Final.))
```

```{r}
GLIS2_vsOthers <- read.csv("DEGs/CBFA2T3GLIS2vsOthers_DEGs.csv")   %>% 
  filter(logFC > 1, 
         Cellular.Compartment_Membrane  != "",
         grepl("^ENSG", gene))
  

dim(GLIS2_vsOthers)
head(GLIS2_vsOthers)
# View(GLIS2_vsOthers)
```


```{r}
GLIS2_vsNBM <- read.csv("DEGs/CBFA2T3GLIS2vsNBM_DEGs.csv") %>% 
  filter(logFC > 1,
         Cellular.Compartment_Membrane  != "",
         grepl("^ENSG", gene))

dim(GLIS2_vsNBM)
```


```{r}
library(rDGIdb)
genes <- c(GLIS2_vsOthers$gene_name, GLIS2_vsNBM$gene_name) %>% 
  unique() #937 unique genes


DGI_Filter <- queryDGIdb(genes = genes,
                             geneCategories = c("CLINICALLY ACTIONABLE"))

```

```{r}
DGI_Final <- detailedResults(DGI_Filter) %>% 
                  group_by(Gene) %>% 
                  #collapse genes with multiple drugs into a single row
                  mutate_at(vars(Drug:PMIDs),
                            ~collapseRows(col = ., uniq = FALSE, sep="; ")) %>% 
                  ungroup()  %>%
                  dplyr::select(-SearchTerm) %>%
                  unique()%>% 
                  
                  mutate_all(~gsub("(;\\s;\\s){1,}|^;\\s|;\\s$","",.)) %>%
  select(gene_name=Gene, Drug.Gene.InteractionDB=Drug, Drug.Gene.InteractionDB_PMIDs=PMIDs)
  


dim(DGI_Final)# 15 hits 

# head(sialicAcid)
```

```{r}
ADC <- ADC_CART %>% 
  filter(Treatment.type == "ADC") %>%
  select(gene_name=Gene.symbol.of.target..Final.,
         ADC_treatment.symbol.or.trade.name=Drug.treatment.symbol.or.trade.name,
         ADC_Full.name.of.drug.treatment=Full.name.of.drug.treatment,
         ADC_Clinical.Trial=If.currently.in.clinical.trials..drug.trial.ID.number) %>% 
  filter(!is.na(ADC_treatment.symbol.or.trade.name)| grepl("^NA$",ADC_treatment.symbol.or.trade.name))



head(ADC)
# table(duplicated(ADC$gene_name)) #OK

CART <- ADC_CART %>% 
  filter(Treatment.type == "CAR T-cell") %>%
  select(gene_name=Gene.symbol.of.target..Final.,
         CART_treatment.symbol.or.trade.name=Drug.treatment.symbol.or.trade.name,
         CART_Full.name.of.drug.treatment=Full.name.of.drug.treatment,
         CART_Clinical.Trial=If.currently.in.clinical.trials..drug.trial.ID.number) %>% 
  filter(!is.na(CART_treatment.symbol.or.trade.name)| grepl("^NA$",CART_treatment.symbol.or.trade.name))



head(CART)
# table(duplicated(CART$gene_name)) #OK


# head(ADC_CART)
```

```{r}
filter(ADC,gene_name== "CD276")
filter(CART,gene_name== "CD276") #looks like a bit of switching happened here. 
```

```{r}
GLIS2_vsOthers.anno <- GLIS2_vsOthers %>% 
  mutate(SialicAcidPathway=case_when(
    gene %in% sialicAcid$Gene.stable.ID ~ "Yes")) %>% 
  mutate(Cell_Adhesion_Gene=ifelse(gene_name %in% unlist(GO.Adhesion), "Yes", NA)) %>%
  
  left_join(.,DGI_Final, by="gene_name") %>%
  left_join(., ADC, by="gene_name") %>% 
  left_join(., CART, by="gene_name")%>%
  
  select(gene,gene_name, SialicAcidPathway,Cell_Adhesion_Gene, everything(),
         -AveExpr, -t, -P.Value,-B,-external_gene_name,
         -geneStableID) 
  



head(GLIS2_vsOthers.anno)
dim(GLIS2_vsOthers.anno)

# write.csv(GLIS2_vsOthers.anno,"DEGs/CBFGLIS/TARGET_AML_CBFGLIS_vs_OtherAML_GRCh38_DEGs_CellMembrane_AdhesionGenes.csv",
#           row.names = FALSE)
```

```{r}
table(GLIS2_vsOthers.anno$SialicAcidPathway)#2

table(GLIS2_vsOthers.anno$Cell_Adhesion_Gene) #146

table(GLIS2_vsOthers.anno$CART_treatment.symbol.or.trade.name) %>% length()
table(GLIS2_vsOthers.anno$ADC_treatment.symbol.or.trade.name) %>% length()
table(GLIS2_vsOthers.anno$Drug.Gene.InteractionDB) %>% length()
```

```{r}
GLIS2_vsNBM.anno <- GLIS2_vsNBM %>% 
  mutate(SialicAcidPathway=case_when(
    gene %in% sialicAcid$Gene.stable.ID ~ "Yes")) %>% 
  mutate(Cell_Adhesion_Gene=ifelse(gene_name %in% unlist(GO.Adhesion), "Yes", NA)) %>%
  
  left_join(.,DGI_Final, by="gene_name") %>%
  left_join(., ADC, by="gene_name") %>% 
  left_join(., CART, by="gene_name")%>%
  
  select(gene,gene_name, SialicAcidPathway, Cell_Adhesion_Gene,everything(),
         -AveExpr, -t, -P.Value,-B,-external_gene_name,
         -geneStableID) 
  



head(GLIS2_vsNBM.anno)
dim(GLIS2_vsNBM.anno)
# 
# write.csv(GLIS2_vsNBM.anno,"DEGs/CBFGLIS/TARGET_AML_CBFGLIS_vs_NBM_GRCh38_DEGs_CellMembrane_AdhesionGenes.csv",
#           row.names = FALSE)
```


```{r}
table(GLIS2_vsNBM.anno$SialicAcidPathway)


table(GLIS2_vsNBM.anno$Cell_Adhesion_Gene) #122

table(GLIS2_vsNBM.anno$Drug.Gene.InteractionDB) %>% length()
table(GLIS2_vsNBM.anno$ADC_treatment.symbol.or.trade.name) %>% length()
table(GLIS2_vsNBM.anno$CART_treatment.symbol.or.trade.name) %>% length()
```



#Session Information 

```{r}
sessionInfo()
```

